---
title: "alphadiv_statsclub"
author: "Alexander Piper"
date: "29/10/2019"
output: html_document
---

## Intro
Steps in analysis:

* Alpha diversity - How many “species” are there? (richness, evenness, or both). Alpha diversity metrics generally operate on a single sample (i.e., within sample diversity). 

* Beta diversity - How similar are pairs of samples? Beta diversity metrics operate on a pair of samples (i.e., between sample diversity).

* Who is there: Taxonomy, differential abundance, association etc

## Load libraries

```{r load packages, message=FALSE, warning=FALSE}
#Set required packages
.cran_packages <- c("tidyverse", "RColorBrewer")
.bioc_packages <- c("phyloseq", "philr")
.github_packages <- c("speedyseq", "breakaway")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}

.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}

.inst <- .github_packages %in% installed.packages()
if(any(!.inst)) {
  devtools::install_github(.github_packages[!.inst])
}

#Load all packages
sapply(c(.cran_packages,.bioc_packages,.github_packages), require, character.only = TRUE)
```

## Get Phyloseq object

```{r load PS}
ps <- readRDS("output/rds/ps_rdp.rds")

ps
```

## Detect and remove outlier Samples

Detecting and potentially removing samples outliers (those samples with underlying data that do not conform to experimental or biological expectations) can be useful for minimizing technical variance. This can be caused by a number of reasons, including low-reads assigned to that sample. In this case we remove all samples below 1000 reads, as these include all samples contributing to lower than usual ASV counts.

```{r remove outliers, warnings=FALSE}
## Remove mocks
rm_mocks <- c("mockA_S51", "MockEven_S193", "Mock_S192", "PCRctrl_S191", "MockStaggered_S194", "PCRctrl_S191", "91_S167")
nsamples(ps)
ps0 <- ps %>% subset_samples(!sample_names(ps) %in% rm_mocks) %>% # Remove mocks

  filter_taxa(function(x) mean(x) > 0, TRUE) # Drop missing taxa from table
nsamples(ps0)

# Plot readcount
threshold <- 1000

gg.asv.boxplot <- merge(as.data.frame(sample_data(ps0)), data.frame("Reads" = sample_sums(ps0)), by = "row.names") %>%
  ggplot(aes(x = Sample.Name, y = Reads, color = Sample.Name, shape = replicated)) +
  geom_boxplot(outlier.colour = "RED", position = position_dodge(width = 0.8)) +
  geom_jitter(size = 2, alpha = 0.6) +
  scale_y_log10() +
  geom_hline(yintercept = threshold, lty = 2) +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90, vjust = 1)
  ) + 
  ylab("log10(Reads") +
  xlab("Psyllid specimen")

print(gg.asv.boxplot)

# Remove all samples under the minimum read threshold
ps1 <- prune_samples(sample_sums(ps0) >= threshold, ps0) #
ps1 <- filter_taxa(ps1, function(x) mean(x) > 0, TRUE) # Drop missing taxa from table
print(paste(nsamples(ps) - nsamples(ps1), " Samples and ", ntaxa(ps) - ntaxa(ps1), " taxa Dropped"))
```


### Merge technical replicates

With only 16 samples replicated, and only on the first 2 sequencing runs i dont think there is a way to explicitly take into account replicate variability, therefore all replicates were merged

```{r merge replicates}
# Merge replicates
ps.merged <- ps1 %>%
  merge_samples(group = "Sample.Name")

# This loses the sample metadata - Need to add it agian
samdf <- read.csv("sample_data/Sample_info.csv", header = TRUE) %>%
  filter(!duplicated(Sample.Name)) %>%
  magrittr::set_rownames(.$Sample.Name) %>%
  dplyr::select(c("SampleID", "Sample.Name", "seqrun", "psyllid_spp", "psyllid_genus", "psyllid_family", "hostplant_spp", "Collection", "Collection.Date"))

sample_data(ps.merged) <- samdf
ps.merged <- filter_taxa(ps.merged, function(x) mean(x) > 0, TRUE) # Drop missing taxa from table
```

## Taxon Filtering

The following R chunk removes taxa not-typically part of a bacterial microbiome analysis.

We also remove all taxa contained within the blank sample from other samples

```{r taxon-cleaning}
get_taxa_unique(ps.merged, "Root")
get_taxa_unique(ps.merged, "Class")

ps.merged # Check the number of taxa prior to removal
ps2 <- ps.merged %>%
  subset_taxa(
    Root == "Bacteria" & # This is probably the only required one
      Family != "mitochondria" &
      Class != "Chloroplast" &
      Phylum != "Cyanobacteria/Chloroplast"
  )
ps2 # Confirm that the taxa were removed
get_taxa_unique(ps2, "Root")
get_taxa_unique(ps2, "Class")
```

# Alpha Diversity

The first question we wish to ask is there a relationship between alpha diversity (richness) of microbial taxa and either psyllid identity or psyllid phylogeny.

While measurement error in microbiome studies affects all analyses of microbiome data, alpha diversity is particularly affected because commonly used estimates of alpha diversity are heavily biased compared to other estimation problems in microbial ecology. One major bias is the high correlation between observed richness and sequencing depth, as we can see in the below plots.

```{r richness diagnostics}
library(breakaway)

plot_richness(ps2)

## Plot observed richness, as a function of sequencing (sampling) depth
gg.richdepth <- data.frame(
  "observed_richness" = (sample_richness(ps2) %>% summary())$estimate,
  "depth" = phyloseq::sample_sums(ps2),
  "type" = ps2 %>% sample_data() %>% get_variable("psyllid_spp")
) %>%
  ggplot(aes(x = depth, y = observed_richness, color = type)) +
  geom_point() +
  geom_smooth(aes(x = depth, y = observed_richness), method = "lm", inherit.aes = FALSE) +
  xlab("Sequencing Depth") +
  ylab("Observed richness") +
  ggtitle("Observed richness by sequencing depth") +
  theme(legend.position = "none")

gg.richrun <- data.frame(
  "observed_richness" = (sample_richness(ps2) %>% summary())$estimate,
  "type" = ps2 %>% sample_data() %>% get_variable("seqrun")
) %>%
  mutate(type = as.factor(type)) %>%
  ggplot(aes(x = type, y = observed_richness, color = type)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter() +
  xlab("Sequencing Run") +
  ylab("Observed richness") +
  ggtitle("Observed richness by sequencing run")

print(gg.richdepth)
print(gg.richrun)
```

## Estimate richness of community, not sample

Due to these biases, instead of using traditional plug-in diversity estimators, which only tell us about the diversity of our samples (observed richness), instead of the actual environment (actual richness), we will use the breakaway package which uses biological replicates to estimate the total diversity in the environment. Breakaway estimates the number of missing species based on the sequence depth and  number of rare taxa in the data, and accounts for experimental noise such as the correlation between observed richness and the sequencing depth.

Model total diversity = a fixed but unknown parameter


```{r estimate richness}
# Model richness for all samples by psyllid spp
ba <- breakaway(ps2)

## Check the specification of the model
betta_pic(summary(ba)$estimate, summary(ba)$error)

ba_estimates <- summary(ba) %>%
  add_column("Sample.Name" = ps2 %>% otu_table() %>% sample_names()) %>%
  left_join(samdf, by = "Sample.Name")

## Plot the richness estimates from each sample 

col <- colorRampPalette(brewer.pal(11, "Spectral"))(75)

pd <- position_dodge(width = 1)

gg.barichnes <- ggplot(ba_estimates, aes(x = psyllid_spp, y = estimate, fill = psyllid_spp)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(aes(colour = psyllid_spp), colour = "black", pch = 21, position = position_jitter(width=0.2, seed = 123)) +
  geom_linerange(aes(ymin = lower, ymax = upper, colour=psyllid_spp), 
                size = 1, position = position_jitter(width=0.2, seed = 123),
                alpha=0.8)  +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0, hjust = 1),
    legend.position = "none"
  ) +
  scale_fill_manual(values = col) +
  scale_colour_manual(values = col) +
  ylab("Richness Estimate") + 
  coord_cartesian(ylim=c(0,1000))

```

# Model richness with psyllid_spp as a fixed effect

Following the initial estimate of diversity and variance using the breakaway package, we use the betta model of Willis, A. and Bunge, J. (2015) which works like regression but takes into account uncertainty in diversirty estimates due to incomplete sampling. Betta model - uses the beta diversity (homogeneity between samples with the same diversity) to feed back and improve the estimates of alpha diversity - via BLUP Framework of linear mixed model.In our case, we have multiple biological replicates. It is very likely that the microbiomes of similar psyllids resemble one another, therefore we can take advantages of common features across multiple samples to reduce the variability in the estimates above

Some of the samples in the initial estimates had large standard errors, indicating that we cannot precisely estimate the total species richness in those sample. In the betta model these samples will be given less weight in the regression than samples with small standard errors on their richness estimates. This is the behaviour we want, because while that sample contains some information, the standard error is telling us that the richness estimate may not be as reliable as other samples' estimates, and we should downweight the importance of that observation relative to observations would smaller standard errors on their richness estimates.



## covariates

betta models species diversity as function of covariates + random variation. Need to choose suitable covariates

```{r covariates}
print(colnames(sample_data(ps2)))
```

## Psyllid_spp as covariate

```{r richness psyllid_spp}
bt_spp <- betta(
  chats = ba_estimates$estimate,
  ses = ba_estimates$error,
  X = model.matrix(~psyllid_spp, data = ba_estimates)
)

bt_spp$table

intercept <- unique(ba_estimates$psyllid_spp[which(!ba_estimates$psyllid_spp %in% bt_sppdf$psyllid_spp)])

bt_sppdf <- as_tibble(bt_spp$table, rownames = "psyllid_spp") %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "psyllid_spp", replacement = "")) %>%
  magrittr::set_colnames(c("psyllid_spp", "Estimates", "SE", "P")) %>%
  mutate(Estimates = case_when(
    psyllid_spp == "(Intercept)" ~ Estimates,
    !psyllid_spp == "(Intercept)" ~ Estimates +
      filter(., str_detect(psyllid_spp, "Intercept")) %>% # Rescale by intercept
      pull(Estimates)
  )) %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "(Intercept)", replacement = intercept)) %>%
  mutate(significant = case_when(
    P < 0.05 ~ TRUE,
    P > 0.05 ~ FALSE
  ))


gg.sppdf_richness <- ggplot(bt_sppdf, aes(x = psyllid_spp, y = Estimates, fill = psyllid_spp)) +
  geom_point(aes(shape = significant, colour = psyllid_spp), size = 3) +
  geom_errorbar(aes(x = psyllid_spp, ymin = Estimates - SE, ymax = Estimates + SE)) +
  geom_hline(yintercept = bt_sppdf %>% filter(psyllid_spp == paste0("(", intercept, ")")) %>% pull(Estimates)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0, hjust = 1),
    legend.position = "none"
  ) +
  scale_fill_manual(values = col) +
  scale_colour_manual(values = col) +
  ylab("Richness Estimate") +
  coord_flip() +
  ggtitle("Estimated taxonomic richness using psyllid_spp as mixed-effect", subtitle = "Triangles indicate significance at P < 0.05")

print(gg.sppdf_richness)
```

## Hostplant_spp as covariate

```{r richness hostplant_spp}
# Model richness with hostplant_spp as a fixed effect
bt_host <- betta(
  chats = ba_estimates$estimate,
  ses = ba_estimates$error,
  X = model.matrix(~hostplant_spp, data = ba_estimates)
)

bt_host$table

bt_hostdf <- as_tibble(bt_spp$table, rownames = "psyllid_spp") %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "psyllid_spp", replacement = "")) %>%
  magrittr::set_colnames(c("psyllid_spp", "Estimates", "SE", "P")) %>%
  mutate(Estimates = case_when(
    psyllid_spp == "(Intercept)" ~ Estimates,
    !psyllid_spp == "(Intercept)" ~ Estimates +
      filter(., str_detect(psyllid_spp, "Intercept")) %>% # Rescale by intercept
      pull(Estimates)
  )) %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "(Intercept)", replacement = intercept)) %>%
  mutate(significant = case_when(
    P < 0.05 ~ TRUE,
    P > 0.05 ~ FALSE
  ))

intercept <- unique(ba_estimates$psyllid_spp[which(!ba_estimates$psyllid_spp %in% bt_hostdf$psyllid_spp)])

gg.hostdf_richness <- ggplot(bt_hostdf, aes(x = psyllid_spp, y = Estimates, fill = psyllid_spp)) +
  geom_point(aes(shape = significant, colour = psyllid_spp), size = 3) +
  geom_errorbar(aes(x = psyllid_spp, ymin = Estimates - SE, ymax = Estimates + SE)) +
  geom_hline(yintercept = bt_hostdf %>% filter(psyllid_spp == paste0("(", intercept, ")")) %>% pull(Estimates)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0, hjust = 1),
    legend.position = "none"
  ) +
  scale_fill_manual(values = col) +
  scale_colour_manual(values = col) +
  ylab("Richness Estimate") +
  coord_flip() +
  ggtitle("Estimated taxonomic richness using hostplant_spp as mixed-effect", subtitle = "Triangles indicate significance at P < 0.05")

print(gg.hostdf_richness)

```

## Mixed + random effects

Model with psyllid_spp as a fixed effect and host_plant spp as a random effect

What else? seqrun? collection date?

```{r with random effects}

bt_sppfixed_hostrandom <- betta_random(
  chats = ba_estimates$estimate,
  ses = ba_estimates$error,
  X = model.matrix(~psyllid_spp, data = ba_estimates), # Set fixed effects
  groups = ba_estimates$hostplant_spp # Set random effects
) 

intercept <- unique(ba_estimates$psyllid_spp[which(!ba_estimates$psyllid_spp %in% bt_randomdf$psyllid_spp)])

bt_randomdf <- as_tibble(bt_sppfixed_hostrandom$table, rownames = "psyllid_spp") %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "psyllid_spp", replacement = "")) %>%
  magrittr::set_colnames(c("psyllid_spp", "Estimates", "SE", "P")) %>%
  mutate(Estimates = case_when(
    psyllid_spp == "(Intercept)" ~ Estimates,
    !psyllid_spp == "(Intercept)" ~ Estimates +
      filter(., str_detect(psyllid_spp, "Intercept")) %>% # Rescale by intercept
      pull(Estimates)
  )) %>%
  mutate(psyllid_spp = str_replace(psyllid_spp, pattern = "(Intercept)", replacement = intercept)) %>%
  mutate(significant = case_when(
    P < 0.05 ~ TRUE,
    P > 0.05 ~ FALSE
  ))

gg.betarichness <- ggplot(bt_randomdf, aes(x = psyllid_spp, y = Estimates, fill = psyllid_spp)) +
  geom_point(aes(shape = significant, colour = psyllid_spp), size = 3) +
  geom_errorbar(aes(x = psyllid_spp, ymin = Estimates - SE, ymax = Estimates + SE)) +
  geom_hline(yintercept = bt_randomdf %>% filter(psyllid_spp == paste0("(", intercept, ")")) %>% pull(Estimates)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0, hjust = 1),
    legend.position = "none"
  ) +
  scale_fill_manual(values = col) +
  scale_colour_manual(values = col) +
  ylab("Richness Estimate") +
  coord_flip() +
  ggtitle("Estimated taxonomic richness using psyllid_spp as mixed-effect, hostplant_spp as random-effect", subtitle = "Triangles indicate significance at P < 0.05")

```


### Sanity check using chao bunge

```{r chao bunge, eval=FALSE}
cb <- chao_bunge(ps2, cutoff=100)

## Check the specification of the model
betta_pic(summary(cb)$estimate, summary(cb)$error)

cb_estimates <- summary(cb) %>%
  add_column("Sample.Name" = ps2 %>% otu_table() %>% sample_names()) %>%
  left_join(samdf, by = "Sample.Name")

## Plot the richness estimates from each sample 

col <- colorRampPalette(brewer.pal(11, "Spectral"))(75)

pd <- position_dodge(width = 1)

gg.barichnes <- ggplot(ba_estimates, aes(x = psyllid_spp, y = estimate, fill = psyllid_spp)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(aes(colour = psyllid_spp), colour = "black", pch = 21, position = position_jitter(width=0.2, seed = 123)) +
  geom_linerange(aes(ymin = lower, ymax = upper, colour=psyllid_spp), 
                size = 1, position = position_jitter(width=0.2, seed = 123),
                alpha=0.8)  +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0, hjust = 1),
    legend.position = "none"
  ) +
  scale_fill_manual(values = col) +
  scale_colour_manual(values = col) +
  ylab("Richness Estimate") + 
  coord_cartesian(ylim=c(0,1000))

```

## Sessioninfo

```{r session info}
sessionInfo()
```
