---
title: "Psyllid microbiome - Statistics"
author: "Alexander Piper"
date: "07/04/2020"
output:
  html_document:
    highlighter: null
    theme: "flatly"
    code_download: true
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    df_print: paged
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Knitr global setup - change eval to true to run code
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, message=FALSE, error=FALSE,fig.show = "hold", fig.keep = "all")
opts_chunk$set(dev = 'png')
```

# Load packages 

```{r install & Load packages} 
#Set required packages
.cran_packages <- c("tidyverse", 
                    "patchwork", 
                    "vegan", 
                    "seqinr",
                    "ape", 
                    "sp",
                    "data.table", 
                    "RColorBrewer",
                    "ggtree", 
                    "castor", 
                    "picante", 
                    "phylosignal", 
                    "adephylo",
                    "TreeDist",
                    "paco",
                    "phytools")
.bioc_packages <- c("dada2",
                    "phyloseq", 
                    "DECIPHER",
                    "Biostrings",
                    "ShortRead", 
                    "philr",
                    "ALDEx2")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}

sapply(c(.cran_packages,.bioc_packages), require, character.only = TRUE)

# Github packages
devtools::install_github("alexpiper/taxreturn")
library(taxreturn)
devtools::install_github("alexpiper/seqateurs")
library(seqateurs)
devtools::install_github("mikemc/speedyseq")
library(speedyseq)
devtools::install_github('ggloor/CoDaSeq/CoDaSeq')
library(CoDaSeq)

#Source internal functions
source('R/BDTT.R')
source('R/phylosymbiosis.R')
source('R/helper_functions.R')
options(stringsAsFactors = FALSE)
```

# Make Phyloseq object

Following taxonomic assignment, the sequence table and taxonomic table are merged into a single phyloseq object alongside the sample info csv.

```{r create PS, eval = FALSE}
seqtab <- readRDS("output/rds/seqtab_curated.rds")
tax <- readRDS("output/rds/tax_IdTaxaExact.rds") 
fitGTR <- readRDS("output/rds/phytree.rds") 

#Load sample information (Put all collection and geographic distances here?)
## ---- samdat ----
samdf <- read_csv("sample_data/Sample_info.csv")  %>%
  dplyr::rename_all(funs( stringr::str_replace_all(., '\\ ', '_')) ) %>%
  mutate(psyllid_spp = psyllid_spp %>% str_remove("\\.") %>%
           str_replace_all(" ", "_") %>%
           str_replace_all("Casuarinicola sp", "Triozid sp")) %>%
  mutate(hostplant_spp = hostplant_spp %>% str_remove("\\.") %>%
           str_replace_all(" ", "_") %>%
           str_to_sentence() %>%
           str_replace_all("Kamahi", "Weinmannia_racemosa") %>%
           str_replace_all("Oleaeria_odorata", "Olearia_odorata"))

write_csv(samdf, "sample_data/Sample_info2.csv")

# Flag replicated samples
samdf <- samdf %>%
  dplyr::mutate(replicated = Sample_Name %in% (samdf %>% group_by(Sample_Name) %>% 
                                                 add_count() %>%
                                           filter(n >1) %>% 
                                           pull(Sample_Name))) %>%
  separate(Collection, into= c("Lat", "Long"), sep=" ", remove = TRUE) %>%
  separate(Lat, paste("lat",c("d","m","s"), sep="_")) %>%
  separate(Long, paste("long",c("d","m","s"), sep="_" )) %>%
  mutate_at(vars(starts_with("lat")), .funs=as.numeric) %>%
  mutate_at(vars(starts_with("long")), .funs=as.numeric) %>%
  mutate(lat=-(lat_d + lat_m/60 + lat_s/60^2),
            long=long_d + long_m/60 + long_s/60^2) %>%
  dplyr::select(c("SampleID", "Sample_Name", "seqrun", "psyllid_spp", "psyllid_genus", "psyllid_family", "hostplant_spp","Collection_Date","replicated", "lat", "long")) %>%
  as.data.frame(stringsAsFactors=FALSE) %>%
  magrittr::set_rownames(.$SampleID) #Collection

## ---- phyloseq ----
ps <- phyloseq(tax_table(tax), sample_data(samdf),
               otu_table(t(seqtab), taxa_are_rows = FALSE),
               phy_tree(fitGTR$tree))

if(nrow(seqtab) > nrow(sample_data(ps))){warning("Warning: All samples not included in phyloseq object, check sample names match the sample metadata")}

#Rename taxa
taxa_names(ps) <- paste0("SV", seq(ntaxa(ps)),"-",tax_table(ps)[,8])

##save phyloseq object
saveRDS(ps, "output/rds/ps.rds")

#Output tables of results
dir.create("output/csv")
dir.create("output/otu_tables/unfiltered/")

##Export raw csv
speedyseq::psmelt(ps) %>%
  write.csv(file = "output/otu_tables/rawdata.csv")

seqateurs::summarise_taxa(ps, "species", "SampleID") %>%
  spread(key="SampleID", value="totalRA") %>%
  write.csv(file = "output/otu_tables/unfiltered/spp_sum.csv")

seqateurs::summarise_taxa(ps, "genus", "SampleID") %>%
  spread(key="SampleID", value="totalRA") %>%
  write.csv(file = "output/otu_tables/unfiltered/gen_sum.csv")
```


## Read in psyllid phylogeny
```{R psyllid phylogeny}
psyllid_tree <- read.tree(text=readLines("sample_data/psyllid_beast_tree.nwk"))

# Match names with alpha diversity
psyllid_tree$tip.label <- psyllid_tree$tip.label %>%
  str_squish() %>%
  str_replace_all(pattern="\\.", replacement=" ") %>%
  str_replace_all(pattern="Acizzia hakae", replacement="Acizzia hakeae") %>%
  str_replace_all(pattern="POLLENISLAND", replacement="POLLEN ISLAND") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiae$", replacement="Ctenarytaina fuchsia A") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeB", replacement="Ctenarytaina fuchsia B") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeC", replacement="Ctenarytaina fuchsia C") %>%
  str_replace_all(pattern="Ctenarytaina clavata", replacement="Ctenarytaina clavata sp ") %>%
  str_replace_all(pattern="Ctenarytaina clavata sp $", replacement="Ctenarytaina clavata sp A") %>%
  str_replace_all(pattern="Ctenarytaina sp$", replacement="Ctenarytaina sp ") %>%
  str_replace_all(pattern="Ctenarytaina spA", replacement="Ctenarytaina sp A") %>%
    str_replace_all(pattern="Ctenarytaina spB", replacement="Ctenarytaina sp B") %>%
  str_replace_all(pattern="Ctenarytaina unknown", replacement="Ctenarytaina insularis") %>%  
  str_replace_all(pattern="Psylla apicalisA", replacement="Psylla frodobagginsi") %>%
  str_replace_all(pattern="Psylla apicalisB", replacement="Psylla apicalis") %>%
  str_replace_all(pattern="carmichaeliae", replacement="carmichaeliae ") %>%
  str_replace_all(pattern="Trioza sp", replacement="Trioza sp ") %>%
  str_replace_all(pattern="Trioza acutaB", replacement="Trioza acuta B") %>%
  str_replace_all(pattern="Trioza gourlay", replacement="Trioza gourlayi") %>%
  str_replace_all(pattern="BRENDAMAY", replacement="BRENDA MAY") %>%
  str_replace_all(pattern="PRICES", replacement="PRICES VALLEY") %>%  
  str_replace_all(pattern="Acizzia sp", replacement="Acizzia errabunda") %>% 
  str_replace_all(pattern=" ", replacement="_") %>%
  trimws(which="right")

# Subset to only those in sample data
psyllid_tree$tip.label[!psyllid_tree$tip.label %in% samdf$psyllid_spp]
pruned.tree <- drop.tip(psyllid_tree, psyllid_tree$tip.label[!psyllid_tree$tip.label %in% samdf$psyllid_spp] )
```

# Filtering
## Remove outlier Samples

Detecting and potentially removing samples outliers (those samples with underlying data that do not conform to experimental or biological expectations) can be useful for minimizing technical variance. This can be caused by a number of reasons, including low-reads assigned to that sample. In this case we remove all samples below 1000 reads, as these include all samples contributing to lower than usual ASV counts.

Rarefaction curves are useful	to	assess	sensitivity	of	sample	size	to	observed	alpha-diversity estimates.
```{r replicate reproducibility}
## Remove mocks
rm_mocks <- c("mockA_S51", "MockEven_S193", "Mock_S192", "PCRctrl_S191", "MockStaggered_S194", "PCRctrl_S191", "91_S167")

#check mocks
ps0 <- ps %>% 
  subset_samples(!sample_names(ps) %in% rm_mocks) %>% #Remove mocks
  filter_taxa(function(x) mean(x) > 0, TRUE) #Drop missing taxa from table 

message((nsamples(ps) - nsamples(ps0)), " outlier samples dropped")

#Plot rarefaction curve
out <- rarecurve(otu_table(ps0), step=100)

rare <- lapply(out, function(x){
  b <- as.data.frame(x)
  b <- data.frame(OTU = b[,1], count = rownames(b))
  b$count <- as.numeric(gsub("N", "",  b$count))
  return(b)
})
names(rare) <- sample_names(ps1)

rare <- map_dfr(rare, function(x){
  z <- data.frame(x)
  return(z)
}, .id = "sample")

# threshold for read removal
threshold = 1000

gg.rare <- ggplot(data = rare)+
  geom_line(aes(x = count, y = OTU, group=sample), alpha=0.5)+
  geom_point(data = rare %>% group_by(sample) %>% top_n(1, count),aes(x = count, y = OTU, colour=(count > threshold))) +
  scale_x_continuous(labels =  scales::scientific_format()) +
  geom_vline(xintercept=threshold, linetype="dashed") +
  labs(colour = "> Read \n Threshold") +
  xlab("Sequence reads") +
  ylab("Observed ASV's")

gg.rare

#Remove all samples under the minimum read threshold 
ps1 <- prune_samples(sample_sums(ps0)>=threshold, ps0) 
ps1 <- filter_taxa(ps1, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
message(nsamples(ps) - nsamples(ps1), " Samples and ", ntaxa(ps) - ntaxa(ps1), " taxa under read threshold Dropped")
```

## Average dissimilarities

Calculate Average dissimilarities between PCR replicates, biological replicates of the same species, hostplants, and different species on same hostplants

```{r Avg Dissimiarity}
# Calculate overall distance
#Rarefy replicates to same read depth
nspec <- specnumber(otu_table(ps1)) # observed number of species
raremax <- min(rowSums(otu_table(ps1)))
rare <- rrarefy(otu_table(ps1), raremax)

dist <- vegan::vegdist(rare, method="bray") %>%
  as.matrix()

# Plot pca
r.pcx <- prcomp(dist)
factoextra::fviz_pca_biplot(r.pcx)

# Reformat to Long format
xy <- t(combn(colnames(dist), 2))
pw <- data.frame(xy, dist = dist[xy]) %>%
  set_colnames(c("SampleID.x", "SampleID.y", "dist")) %>%
  left_join(samdf %>% 
              dplyr::select(SampleID, Sample_Name, psyllid_spp, psyllid_genus, hostplant_spp, replicated) %>%
              dplyr::rename(SampleID.x = SampleID), by="SampleID.x")%>%
  left_join(samdf %>% 
              dplyr::select(SampleID, Sample_Name, psyllid_spp, psyllid_genus, hostplant_spp, replicated) %>%
              dplyr::rename(SampleID.y = SampleID), by="SampleID.y")


# Distance between replicates of same sample
repdist <- pw %>% filter(Sample_Name.x == Sample_Name.y)

# Distance between samples of same species
sppdist <- pw %>% filter(psyllid_spp.x == psyllid_spp.y)

# Distance between psyllids on same hostplat
#      psyllopsis fraxini 94- 107 - 113
# psyllopsis fraxinicola 93 - 106 - 112
# 93+94 were the same tree, 106+107 on the same tree and 112+113 were the same treee
hostplantdist <- pw %>% filter(psyllid_spp.x == psyllid_spp.y)

dist <- ps1 %>% 
    subset_samples(Sample_Name %in%  c("93", "94","106", "107","112", "113")) %>%
    filter_taxa(function(x) mean(x) > 0, TRUE) %>%
    otu_table()  %>%
    vegan::vegdist(method="jaccard")

factoextra::fviz_pca(prcomp(dist))

# Otther set
#psylla apicalis 200 - 201
#psylla frodobagginsi 200 - 201 #in this case they are named after their tree - ie 200 and 200 were same tree, i shoul drename thes ein some way

dist2 <- ps1 %>% 
    subset_samples(Sample_Name %in%  c("200", "94","106", "107","112", "113")) %>%
    filter_taxa(function(x) mean(x) > 0, TRUE) %>%
    otu_table()  %>%
    vegan::vegdist(method="jaccard")

# could i do some cluster analysis here?

# Or is adonis besst
#adonis(~ psyllid_spp, method = "euclidean", data = metadata))
```

### Merge technical replicates

With only 16 samples replicated, and only on the first 2 sequencing runs i dont think there is a way to explicitly take into account replicate variability, therefore all replicates were merged

```{r replicates}
# Merge replicates
ps.merged <- ps1 %>%
    merge_samples(group = "Sample_Name")

#This loses the sample metadata - Need to add it agian
sample_data(ps.merged) <- sample_data(ps1) %>%
  filter(!duplicated(Sample_Name)) %>%
  magrittr::set_rownames(.$Sample_Name)

#Drop missing taxa from table
ps.merged <- filter_taxa(ps.merged, function(x) mean(x) > 0, TRUE)
```

## Prevalence assesment

View prevalence of different phyla
```{r prevalence-assessment}
# Calculate taxon prevalence across the data set
prevdf <- apply(X = otu_table(ps.merged), MARGIN = ifelse(taxa_are_rows(ps.merged), yes = 1, no = 2), FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf - change this to tidyverse code
prevdf <- data.frame(Prevalence = prevdf, TotalAbundance = taxa_sums(ps.merged), tax_table(ps.merged))

#Prevalence plot
gg.prev <- subset(prevdf, phylum %in% get_taxa_unique(ps0, "phylum")) %>%
  ggplot(aes(TotalAbundance, Prevalence / nsamples(ps.merged),color=family)) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~phylum) +
  theme(legend.position="none") +
  ggtitle("Phylum Prevalence in All Samples\nColored by Family")

gg.prev
```

## Taxon Filtering

Remove all non-bacterial taxa, mitochondria, chloroplast and cyanobacteria.
We also remove all taxa contained within the blank sample from other samples? - Check these

```{r taxon-cleaning}
get_taxa_unique(ps.merged, "domain")

ps.merged # Check the number of taxa prior to removal
ps2 <- ps.merged %>%
  subset_taxa(
    domain == "Bacteria" & #This is probably the only required one
    family  != "Mitochondria" &
    order   != "Chloroplast" &
    phylum != "Cyanobacteria"
  )
ps2 # Confirm that the taxa were removed
get_taxa_unique(ps2, "phylum")
get_taxa_unique(ps2, "class")
get_taxa_unique(ps2, "family")
```

# Alpha diversity 

### Summarise taxonomic assignment

```{r sum taxa}
#Fraction of reads assigned to each taxonomic rank
sum_reads <- speedyseq::psmelt(ps) %>%
  gather("Rank","Name",rank_names(ps)) %>%
  group_by(Rank) %>% 
  mutate(Name = replace(Name, str_detect(Name, "__"),NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  dplyr::summarise(Reads_classified = sum(Abundance * !is.na(Name))) %>%
  mutate(Frac_reads = Reads_classified / sum(sample_sums(ps))) %>%
  mutate(Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank)

#Fraction of ASV's assigned to each taxonomic rank
sum_otu <- tax_table(ps) %>%
  as("matrix") %>%
  as_tibble(rownames="OTU") %>%
  gather("Rank","Name",rank_names(ps)) %>%
  group_by(Rank) %>%
  mutate(Name = replace(Name, str_detect(Name, "__"), NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  dplyr::summarise(OTUs_classified = sum(!is.na(Name))) %>%
  mutate(Frac_OTUs = OTUs_classified / ntaxa(ps)) %>%
  mutate(Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank)

print(sum_reads)
print(sum_otu)
```

## Alpha diversity metrics

Test this with rarefied samples to see the robustness to sequencing depth variation - AS this could be driven by just seq depth
```{r alpha diversity}
phy <- phyloseq::phy_tree(ps2)
phydist <- cophenetic(phy)
comm <- as(phyloseq::otu_table(ps2), "matrix") 

# Get richness measures
richness <- phyloseq::estimate_richness(ps2, measures=c("Shannon")) %>%
  rownames_to_column("Sample_Name") %>%
  mutate(Sample_Name = Sample_Name %>% 
           str_remove("^X") %>%
           str_replace_all("\\.", " "))

#Set number of randomisations for calculating significance
randoms <- 10

# Calculate Faith's PD-index & Species richness - with Standard errors
sespd <- picante::ses.pd(comm, phy, null.model = "taxa.labels", include.root = F, runs = randoms)

# Join together
div_table <- sespd %>%
  rownames_to_column("Sample_Name") %>%
  dplyr::select(Sample_Name, ntaxa, pd.obs.z) %>%
  dplyr::rename(pd = pd.obs.z, alpha = ntaxa) %>%
  left_join(richness, by="Sample_Name") %>%
  left_join(sample_data(ps2) %>% 
              as.data.frame() %>%
              filter(!duplicated(Sample_Name)) %>%
              dplyr::select(Sample_Name, psyllid_spp),
            by = "Sample_Name") 
# Summarise means
div_table %>%
  summarise_if(is.numeric, mean)
```

## Association with phylogeny 

Here we calculate the phylogenetic signal in the alpha diversity and community assembly metrics, or the extent to which these values are statistically related to phylogeny.

see supplementary 15 - https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-019-10191-3/MediaObjects/41467_2019_10191_MOESM1_ESM.pdf

see : https://doi.org/10.1111/j.2041-210X.2012.00196.x for deciding indices
```{r Phylosignal}
dat <- div_table  %>%
  filter(psyllid_spp %in% pruned.tree$tip.label) %>% #Subset to common 
  group_by(psyllid_spp) %>%
  dplyr::select(-Sample_Name) %>%
  summarise_all(mean) %>%
  arrange(match(psyllid_spp, pruned.tree$tip.label)) %>%
  as.data.frame() %>%
  magrittr::set_rownames(.$psyllid_spp) %>%
  dplyr::select(-psyllid_spp)

# Add positive and negative controls
dat$random <- rnorm(length(dat$alpha), sd = 10) #Random association
dat$bm <- rTraitCont(pruned.tree) #Brownian motion

# Make phylosignal object and measure signal between univariate traits.
p4d <- phylobase::phylo4d(pruned.tree, dat)	
signal <- phylosignal::phyloSignal(p4d = p4d, methods = c("I", "Lambda", "K"), reps = 999)

# Plots
barplot.phylo4d(p4d, tree.type = "phylo", tree.ladderize = TRUE)

# Locate signal
lipa <- lipaMoran(p4d, reps=999)
lipa.p4d <- lipaMoran(p4d, as.p4d = TRUE, reps=999)

barplot.phylo4d(p4d, bar.col=(lipa$p.value < 0.05) + 1, center = FALSE , scale = FALSE)

barplot.phylo4d(lipa.p4d, bar.col = (lipa$p.value < 0.05) + 1, center = FALSE, scale = FALSE)

## Should be able to make these in ggtree if necessary 
```	

# Beta diversity

## PCA plot
```{r}


```

## Barplot

Use the means and standard errors? of the beta diversities by sample

Merge species, arrange microbiome by phylogeny, try the mantel tests on the merged data as well (but the seperate is probs more statistically better)

RESULTS: while there were significant differences between psyllid species there were no significant associations with phylogeny
```{r Figure 1}
ps3_bar <- #subset_taxa(ps3, !Family == "Enterobacteriaceae") %>%
            ps3 %>%
  speedyseq::tax_glom(taxrank = "order") %>%           # agglomerate at Order level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  speedyseq::psmelt() %>%
  mutate(plotlabel = phylum) %>%
  mutate(plotlabel = case_when(
    Abundance >= 0.01 & phylum=="Proteobacteria" ~ paste0(phylum, " - ", order), # Change this to whatever taxrank we want
    Abundance >= 0.01 & !phylum=="Proteobacteria"~ phylum ,
    Abundance < 0.01 ~ "NA"
    )) %>%
  dplyr::na_if("NA") %>%
  dplyr::select(Sample, plotlabel, phylum, order, Abundance)

# Plot them together
library(ggtree)
library(treeio)
library(ggstance)
p <- ggtree(pruned.tree) + geom_tiplab(align=TRUE) + geom_nodelab(geom='label') +
    scale_x_continuous(expand=c(0, 0.1)) + theme_tree2()


Fig1 <- facet_plot(p, panel='Relative Abundance', data = ps3_bar %>% mutate(id = Sample), geom=ggstance::geom_barh,
                 mapping=aes(x = Abundance, fill = plotlabel), stat="identity")  + 
  scale_fill_manual(values=colorRampPalette(brewer.pal(11, "Set1"))(length(unique(ps3_bar$plotlabel))-1), na.value="grey")


Fig1 <- facet_plot(Fig1, panel='Species Richness', data = bt_randomdf %>% mutate(id = psyllid_spp),
                          geom=ggstance::geom_pointrangeh,
                           mapping=aes(x=Estimates, xmin=Estimates-SE, xmax=Estimates+SE),
                           stat="identity", size=0.5, colour="#696969") +
  theme_classic2()   +
  theme(legend.position = "bottom",
  panel.grid.major.x = element_line(colour="grey92", size=0.5, linetype="dashed"),
  strip.background = element_rect(fill = "grey92", 
                colour = "black", size = 1)
  )

## Collection_hist
gg.spp <- as_tibble(sample_data(ps2)) %>%
  dplyr::rename(label = psyllid_spp) %>%
  group_by(label) %>%
  summarise(n_species = n()) %>%
  left_join(p$data %>%
  filter(isTip) %>% select(c(label, y))) %>%
  filter(!is.na(y)) %>%
  ggplot(aes(x=factor(.$label, levels=.$label[order(.$y)]), y=1, fill=n_species)) +
  geom_tile() +
  coord_flip() + theme_void() +
  scale_fill_distiller(palette = "Reds", direction=1)

Fig1 + gg.spp 
  
# Change relative grid sizes
library(grid)
library(gtable)

gt = ggplot_gtable(ggplot_build(Fig1))
gt$layout$l[grep('panel-3', gt$layout$name)] # you want to find the column specific to panel-2

gt$widths[9] = 0.3*gt$widths[9] # reduce width of column 9
grid.draw(gt) # plot with grid draw
```


## Phylosymbiosis

### Prepare distance matrices
```{r prepare distances}
## Psyllid phylogeny cophenetic distance matrix
phylo.dist <- cophenetic(pruned.tree) %>%
  as.data.frame() %>%
  rownames_to_column("psyllid_spp.x") %>%
  pivot_longer(cols=-psyllid_spp.x,
               names_to="psyllid_spp.y",
               values_to = "dist") %>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, psyllid_spp) %>%
              dplyr::rename(psyllid_spp.x = psyllid_spp, Sample_Name.x = Sample_Name),
            by="psyllid_spp.x")%>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, psyllid_spp) %>%
              dplyr::rename(psyllid_spp.y = psyllid_spp, Sample_Name.y = Sample_Name),
            by="psyllid_spp.y") %>%
  filter(!is.na(dist))%>%
  dplyr::select(-psyllid_spp.y, -psyllid_spp.x) %>%
  pivot_wider(names_from = Sample_Name.y, values_from = dist) %>%
  magrittr::set_rownames(.$Sample_Name.x) %>%
  dplyr::select(-Sample_Name.x)%>%
  as.matrix()

# New plant  distance matrix
#library(brranching)
# Do this replacement at start
#plants <- samdf  %>%
#  dplyr::select(hostplant_spp)%>%
#  unique() %>%
#  mutate(branching = brranching::phylomatic_names(.$hostplant_spp, 
#                                                format = "isubmit", 
#                                                db = "ncbi"))
#
## Get tree
#plant.tree <- brranching::phylomatic(plants$branching, taxnames = TRUE)
#
## add branch lengths using the wikstrom ages file
#plant.tree <- brranching::rbladj(plant.tree, "wikstrom.ages")
#plant.tree$tip.label <- str_to_sentence(plant.tree$tip.label)
#write.tree(plant.tree, file="sample_data/plant_tree.nwk")

plant.tree <- read.tree("sample_data/plant_tree.nwk")

plant.dist <- cophenetic(plant.tree) %>%
  as.data.frame() %>%
  rownames_to_column("hostplant_spp.x") %>%
  pivot_longer(cols=-hostplant_spp.x,
               names_to="hostplant_spp.y",
               values_to = "dist") %>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, hostplant_spp) %>%
              dplyr::rename(hostplant_spp.x = hostplant_spp, Sample_Name.x = Sample_Name),
            by="hostplant_spp.x")%>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, hostplant_spp) %>%
              dplyr::rename(hostplant_spp.y = hostplant_spp, Sample_Name.y = Sample_Name),
            by="hostplant_spp.y") %>%
  filter(!is.na(dist))%>%
  dplyr::select(-hostplant_spp.y, -hostplant_spp.x) %>%
  pivot_wider(names_from = Sample_Name.y, values_from = dist) %>%
  magrittr::set_rownames(.$Sample_Name.x) %>%
  dplyr::select(-Sample_Name.x)%>%
  as.matrix()

# Spatial distance matrix
envData <- sample_data(ps2) %>%
  as_data_frame() %>%
  dplyr::select(Sample_Name, lat, long) %>%
  tibble::column_to_rownames("Sample_Name") %>%
  drop_na()

spat.dist <- spDists(as.matrix(envData), longlat=TRUE) %>%
  as.data.frame() %>%
  magrittr::set_rownames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  magrittr::set_colnames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  column_to_rownames("SampleID") %>%
  as.matrix()
```

### Whole dataset

But also see how this paper has done it with PACO https://www.nature.com/articles/s41467-019-10191-3.pdf Using random subsampling? Then pull out the most associated also file:///C:/Users/ap0y/Dropbox/workprojects/Papers/10.1002_ecy.1955.pdf

```{r phylosymbiosis}
dir.create("output/phylosymbiosis")
# Get OTU tables
otutab <- otu_table(ps2)
#Impute zeroes for compositional distances
otutab_n0 <- as.matrix(zCompositions::cmultRepl(otutab, method="CZM", output="prop"))
#Root phylogenetic tree
phy_tree(ps2) <- multi2di(phy_tree(ps2))

#Calculate different distance metrics
metrics <- c("Bray", "Jaccard", "Aitchison","Philr", "Unifrac", "WUnifrac") 
distlist <- vector("list", length=length(metrics))
names(distlist) <- metrics

distlist$Jaccard <- as.matrix(vegdist(otutab, method="jac",binary = T))
distlist$Bray <- as.matrix(vegdist(otutab, method="bray"))
distlist$Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_n0), method="euclidean"))
distlist$Philr <- as.matrix(vegdist(philr::philr(otutab_n0, phy_tree(ps2),
                                                part.weights='enorm.x.gm.counts',
                                                ilr.weights='blw.sqrt'), method="euclidean"))
distlist$Unifrac <- as.matrix(phyloseq::UniFrac(ps2, weighted=FALSE, parallel = TRUE))
distlist$WUnifrac <- as.matrix(phyloseq::UniFrac(ps2, weighted=TRUE, parallel = TRUE))

# Adonis test
metadata <- sample_data(ps2) %>%
  as_data_frame()
adonis_results <- distlist %>%
  purrr::map(function(x) {
    bind_rows(
    broom::tidy(adonis(x~psyllid_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1),
    broom::tidy(adonis(x~hostplant_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1)
    )
})  %>%
  bind_rows(.id="dist")

write_csv(adonis_results, "output/phylosymbiosis/adonis_fulldata.csv")

# Matrix correlations
#only use samples present in all
samples <- Reduce(intersect, list(rownames(otu_table(ps2)), colnames(phylo.dist), colnames(plant.dist), colnames(spat.dist)))

# Mantel test
mantel_results <- distlist %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               samples=samples, type="mantel")
  }) %>%
  bind_rows(.id="dist") 

write_csv(mantel_results, "output/phylosymbiosis/mantel_fulldata.csv")

# Partial Mantel Test
pmantel_results <- distlist %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               samples=samples, type="partial")
  }) %>%
  bind_rows(.id="dist") 

write_csv(pmantel_results, "output/phylosymbiosis/pmantel_fulldata.csv")

# Topological comparison metrics
topo_results <- distlist %>%
  purrr::map(function(x){
    run_tree_comparison(x, dists=c("phylo.dist", "plant.dist", "spat.dist"), #Change these to matrices
               samples=samples,
               clust.method="average",
               measure="all",
               runs=100)
  }) %>%
  bind_rows(.id="metric") 
write_csv(topo_results, "output/phylosymbiosis/topo_fulldata.csv")

## Visualise what these look like
#tree1  <- as.phylo(hclust(as.dist(phylo.dist[samples,samples]), method = "ward.D2"))
#tree2 <- as.phylo(hclust(as.dist(distlist$Aitchison[samples,samples]), method = "ward.D2"))
#VisualizeMatching(RobinsonFoulds,tree1, tree2)

```

Potentially just drop the topo comparison in favor of PACO

### Without Gammaproteobacteria

Probably dont need this bit if we are lookign at contribution to cophylogeny through paco
```{r phylosymbiosis no entero}
## TRY WITHOUT Gammaproteobacteria
ps2_subset <- ps2 %>%
 subset_taxa(class != "Gammaproteobacteria") %>% #is this working?
 filter_taxa(function(x) mean(x) > 0, TRUE)#Drop missing taxa from table
ps2_subset <- prune_samples(sample_sums(ps2_subset) >0 , ps2_subset)
message(nsamples(ps2_subset) - nsamples(ps2_subset), " Samples and ", ntaxa(ps2) - ntaxa(ps2_subset), " taxa under read threshold Dropped")

# Get OTU tables
otutab <- otu_table(ps2_subset)
#Impute zeroes for compositional distances
otutab_n0 <- as.matrix(zCompositions::cmultRepl(otutab, method="CZM", output="prop", suppress.print=quiet))
#Root phylogenetic tree
phy_tree(ps2_subset) <- multi2di(phy_tree(ps2_subset))

#Calculate different distance metrics
metrics <- c("Bray", "Jaccard", "Aitchison", "Unifrac", "WUnifrac") # "Philr"
distlist <- vector("list", length=length(metrics))
names(distlist) <- metrics

distlist$Jaccard <- as.matrix(vegdist(otutab, method="jac",binary = T))
distlist$Bray <- as.matrix(vegdist(otutab, method="bray"))
distlist$Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_n0), method="euclidean"))
distlist$Philr <- as.matrix(vegdist(philr::philr(otutab_n0, phy_tree(ps2),
                                                part.weights='enorm.x.gm.counts',
                                                ilr.weights='blw.sqrt'), method="euclidean"))
distlist$Unifrac <- as.matrix(phyloseq::UniFrac(ps2, weighted=FALSE, parallel = TRUE))
distlist$WUnifrac <- as.matrix(phyloseq::UniFrac(ps2, weighted=TRUE, parallel = TRUE))

# Adonis test
metadata <- sample_data(ps2_subset) %>%
  as_data_frame()
adonis_results <- distlist %>%
  purrr::map(function(x) {
    bind_rows(
    broom::tidy(adonis(x~psyllid_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1),
    broom::tidy(adonis(x~hostplant_spp, method="euclidean", data=metadata)$aov.tab)%>% dplyr::slice(1)
    )
})  %>%
  bind_rows(.id="dist")

write_csv(adonis_results, "output/phylosymbiosis/adonis_subset.csv")

# Matrix correlations
#only use samples present in all
samples <- Reduce(intersect, list(rownames(otu_table(ps2)), colnames(phylo.dist), colnames(plant.dist), colnames(spat.dist)))

# Mantel test
mantel_results <- distlist %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               samples=samples, type="mantel")
  }) %>%
  bind_rows(.id="dist") 
write_csv(mantel_results, "output/phylosymbiosis/mantel_subset.csv")

# Partial Mantel Test
pmantel_results <- distlist %>%
  map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               samples=samples, type="partial")
  }) %>%
  bind_rows(.id="dist") 
write_csv(pmantel_results, "output/phylosymbiosis/pmantel_subset.csv")

# Topological comparison metrics
topo_results <- distlist %>%
  purrr::map(function(x){
    run_tree_comparison(x, dists=c("phylo.dist", "plant.dist", "spat.dist"), #Change these to matrices
               samples=samples,
               clust.method="average",
               measure="all",
               runs=100)
  }) %>%
  bind_rows(.id="metric") 
write_csv(topo_results, "output/phylosymbiosis/topo_subset.csv")
```

## BDTT

```{r timeslice}
library(TreeDist)
#Look at depth categories in tree
hist(get_all_node_depths(phy_tree(ps2)))

#Set slices
slices=seq(0,1.6,0.1)

treelist <- vector("list", length=length(slices))
for (i in 1:length(slices)){
  newps <- speedyseq::tree_glom(ps2, resolution = slices[i])
  treelist[[i]] <- phy_tree(newps)
}
class(treelist) <- "multiPhylo"
names(treelist) <- paste0("Timeslice = ",slices)
gg.tree <- ggtree(treelist) + facet_wrap(~.id, scale="free", nrow=1) + theme_tree2()

Betas <- BDTT(ps2, slices, metrics=c("Bray", "Aitchison", "Philr"), zeroes="Impute", cores=1, quiet=FALSE) #"Unifrac", "WUnifrac"

# Matrix correlations
#only use samples present in all
samples <- Reduce(intersect, list(rownames(otu_table(ps2)), colnames(phylo.dist), colnames(coi.dist), colnames(plant.dist),colnames(newplant.dist), colnames(spat.dist)))

# Partial Mantel
BTDD_pm <- Betas %>%
  purrr::map_depth(.depth=2, function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               samples=samples, type="partial")
  }) %>%
  purrr::map(bind_rows,.id="dist") %>%
  bind_rows(.id="slice")

gg.pmantel <- BTDD_pm %>%
  mutate(Comparison = paste0("microbe~",dist1,"~",dist2)) %>%
  ggplot(aes(y=statistic, x=slice, colour=Comparison, group=factor(Comparison)))+ 
  geom_line(linetype="dashed") +
  geom_line(data=BTDD_pm %>% mutate(Comparison = paste0("microbe~",dist1,"~",dist2)) %>% mutate(statistic = case_when (signif > 0.01 ~ as.double(NA), signif < 0.01 ~ statistic)), linetype="solid") +
  geom_hline(yintercept=0) +
  facet_wrap(~dist) +
     #theme_classic2()   +
     theme(legend.position = "right",
           panel.grid.major.y = element_line(colour="grey", size=0.5, linetype="dashed"),
           strip.background = element_rect(fill = "grey92", 
                                           colour = "black", size = 1)
     )  + scale_color_brewer(palette="Paired")

gg.pmantel
```

# Cophylogeny
Due to the association with close branches we looked for patterns of cophylogeny

ggtree tanglegrams: https://stackoverflow.com/questions/57161083/coloring-lines-in-tanglegram-based-on-position-of-nodes

## Whole dataset - PACO

```{r PACO}
# Merge species for beta diversity
ps.sppmerged <- ps2 %>%
    merge_samples(group = "psyllid_spp", fun=mean)

#This loses the sample metadata - Need to add it agian
sample_data(ps.sppmerged) <- sample_data(ps2) %>%
  filter(!duplicated(psyllid_spp)) %>%
  magrittr::set_rownames(.$psyllid_spp)

#Rename taxa - only keep first 30 characters
#taxa_names(ps3) <- substr(paste0("SV", seq(ntaxa(ps3)),"-",tax_table(ps3)[,7]), 1,30)

#Prepare co-occurance matrix
coocur <- ps3 %>%
    subset_samples(!psyllid_spp == "Atmetocranium_myersi") %>%
    otu_table %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0))

colnames(coocur) <- colnames(coocur) %>%
           str_replace_all(" |-", "_")
rownames(coocur) <- rownames(coocur) %>%
           str_replace_all(" |-", "_")

pruned.tree2 <- drop.tip(pruned.tree, "Atmetocranium_myersi" )

# H cophenetic distance
h_tree <- pruned.tree2
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- cophenetic(h_tree)

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- cophenetic(s_tree)

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='cailliez')

# plot Principal coordinates
D$H_PCo <-  neg_to_pos_eig(D$H_PCo)
D$P_PCo <-  neg_to_pos_eig(D$P_PCo)

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) #Symetric - is one meant to track the evolution of another?

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using jacknifing
#paco_run <- paco_links(paco_run)

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=FALSE) 

# Extract the goods for ggtree
# psyllid_tree
p1 <- ggtree(obj[["trees"]][[1]], ladderize=FALSE) + geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(res %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(values)) %>%
    dplyr::rename(label = Sample_Name)
    )
p1 <- p1 %<+% weights_p1  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# OTU tree
p2 <- ggtree(obj[["trees"]][[2]], ladderize=FALSE) #+ geom_tiplab()
weights_p2 <- p2$data %>%
  left_join(res %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)) %>%
    dplyr::rename(label = OTU)
    ) 

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(res %>% 
              dplyr::select(Sample_Name, OTU, values) %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

gg.tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup %>%
  #mutate(values = as.numeric((1-values)/100)) %>%
  ggplot(aes(x=tree, y=y, group=assoc, colour=values)) +
    geom_line(alpha=0.1) +  
    scale_color_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = c(0,0)) + 
    theme_void() +
    theme(legend.position = "bottom")

whole_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
whole_tanglegram
```

## Carsonella cophylogeny

Definitely need to do the top carsonella, but need some flexibility in keeping the related ones, so top - 20% abundance? 

```{r subsetting}
#Flag top abundance carsonella epr sample
top_carson <- ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  filter(genus=="Candidatus Carsonella") %>%
  group_by(Sample) %>%  
  filter(Abundance > 0) %>%
  top_n(1, wt=Abundance) %>%
  mutate(top = TRUE) 

coocur <- ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  #filter(genus=="Candidatus Carsonella") %>%
  left_join(top_carson) %>%
  filter(genus=="Candidatus Carsonella", top==TRUE) %>%
  dplyr::select(OTU, psyllid_spp, SampleID, Abundance) %>%
  mutate(OTU = OTU %>%
           str_replace_all(" |-", "_")) %>%
  dplyr::group_by(OTU, psyllid_spp) %>%
    summarise(Abundance = sum(Abundance)) %>%
  pivot_wider(id_cols = psyllid_spp,
              names_from = OTU,
              values_from=Abundance,
              values_fill = list(Abundance = 0))  %>%
  column_to_rownames("psyllid_spp") %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0)) 

# H cophenetic distance
h_tree <- pruned.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- cophenetic(h_tree)

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- cophenetic(s_tree)

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='cailliez')

# plot Principal coordinates
D$H_PCo <-  neg_to_pos_eig(D$H_PCo)
D$P_PCo <-  neg_to_pos_eig(D$P_PCo)

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) #Symetric - is one meant to track the evolution of another?

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using jacknifing
paco_run <- paco_links(paco_run)

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# psyllid_tree
p1 <- ggtree(obj[["trees"]][[1]], ladderize=FALSE) + geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(res %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(values)) %>%
    dplyr::rename(label = Sample_Name)
    )
p1 <- p1 %<+% weights_p1  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# OTU tree
p2 <- ggtree(obj[["trees"]][[2]], ladderize=FALSE) + geom_tiplab()
weights_p2 <- p2$data %>%
  left_join(res %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)) %>%
    dplyr::rename(label = OTU)
    ) 

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(res %>% 
              dplyr::select(Sample_Name, OTU, values) %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

gg.tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup %>%
  #mutate(values = as.numeric((1-values)/100)) %>%
  ggplot(aes(x=tree, y=y, group=assoc, colour=values)) +
    geom_line(alpha=1) +  
    scale_color_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = c(0,0)) + 
    theme_void() +
    theme(legend.position = "bottom")

carsonella_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
carsonella_tanglegram
```

## Psyllid hostplant cophylogeny

Tanglegram of all plants and  psyllids!
```{r hostplant-psyllid}
plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.tree  <- drop.tip(plant.tree , "Sophora_microphylla_-kowhai" )

#Prepare co-occurance matrix
coocur <- samdf %>%
  as_tibble() %>%
  dplyr::select(psyllid_spp, hostplant_spp) %>%
  filter(!is.na(psyllid_spp)) %>%
  unique() %>%
  mutate(psyllid_spp = psyllid_spp %>%
           str_replace_all(" |-", "_"),
         hostplant_spp = hostplant_spp %>%
           str_replace_all(" |-", "_"),
         presence = 1
         ) %>%
  pivot_wider(names_from = "hostplant_spp",
              values_from="presence",
              values_fill = list(presence = 0)) %>%
  column_to_rownames("psyllid_spp") %>%
  t()

# H cophenetic distance
h_tree <- plant.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- cophenetic(h_tree)

# P cophenetic distance
s_tree <- pruned.tree
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- cophenetic(s_tree)

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]
# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='cailliez') 

# plot Principal coordinates
D$H_PCo <-  neg_to_pos_eig(D$H_PCo)
D$P_PCo <-  neg_to_pos_eig(D$P_PCo)

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) #Symetric - is one meant to track the evolution of another?

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using leave-one-out jacknifing
paco_run <- paco_links(paco_run, .parallel = TRUE)

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge")

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  #facet_wrap(~psyllid_genus) +
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# plant tree
p1 <- ggtree(obj[["trees"]][[1]], ladderize=FALSE) + geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(res %>%
  group_by(hostplant_spp) %>%
  summarise(values = mean(values)) %>%
    dplyr::rename(label = hostplant_spp)
    )
p1 <- p1 %<+% weights_p1  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# psyllid_tree
p2 <- ggtree(obj[["trees"]][[2]], ladderize=FALSE) + geom_tiplab()
weights_p2 <- p2$data %>%
  left_join(res %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(values)) %>%
    dplyr::rename(label = psyllid_spp)
    ) 

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(res %>% 
              dplyr::select(hostplant_spp, psyllid_spp, values) %>% 
              dplyr::rename(label.x = hostplant_spp, label.y = psyllid_spp)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

gg.tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup %>%
  #mutate(values = as.numeric((1-values)/100)) %>%
  ggplot(aes(x=tree, y=y, group=assoc, colour=values)) +
    geom_line(alpha=1) +  
    scale_color_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = c(0,0)) + 
    theme_void() +
    theme(legend.position = "bottom")

plant_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
plant_tanglegram 
```


## Trioza hostplants and trioza psyllids

```{r trioza hostplant psyllid}
plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.tree  <- drop.tip(plant.tree , "Sophora_microphylla_-kowhai" )

#Prepare co-occurance matrix
coocur <- samdf %>%
  as_tibble() %>%
  filter(psyllid_genus == "Trioza") %>%
  dplyr::select(psyllid_spp, hostplant_spp) %>%
  filter(!is.na(psyllid_spp)) %>%
  unique() %>%
  mutate(psyllid_spp = psyllid_spp %>%
           str_replace_all(" |-", "_"),
         hostplant_spp = hostplant_spp %>%
           str_replace_all(" |-", "_"),
         presence = 1
         ) %>%
  pivot_wider(names_from = "hostplant_spp",
              values_from="presence",
              values_fill = list(presence = 0)) %>%
  column_to_rownames("psyllid_spp") %>%
  t()

# H cophenetic distance
h_tree <- plant.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- cophenetic(h_tree)

# P cophenetic distance
s_tree <- pruned.tree
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- cophenetic(s_tree)

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]
# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='cailliez') 

# plot Principal coordinates
D$H_PCo <-  neg_to_pos_eig(D$H_PCo)
D$P_PCo <-  neg_to_pos_eig(D$P_PCo)

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) #Symetric - is one meant to track the evolution of another?

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using jacknifing
paco_run <- paco_links(paco_run)

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge")

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# plant tree
p1 <- ggtree(obj[["trees"]][[1]], ladderize=FALSE) + geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(res %>%
  group_by(hostplant_spp) %>%
  summarise(values = mean(values)) %>%
    dplyr::rename(label = hostplant_spp)
    )
p1 <- p1 %<+% weights_p1  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# psyllid_tree
p2 <- ggtree(obj[["trees"]][[2]], ladderize=FALSE) + geom_tiplab()
weights_p2 <- p2$data %>%
  left_join(res %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(values)) %>%
    dplyr::rename(label = psyllid_spp)
    ) 

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=values)) +  scale_color_gradient(low="steelblue", high="darkorange1")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(res %>% 
              dplyr::select(hostplant_spp, psyllid_spp, values) %>% 
              dplyr::rename(label.x = hostplant_spp, label.y = psyllid_spp)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

gg.tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup %>%
  #mutate(values = as.numeric((1-values)/100)) %>%
  ggplot(aes(x=tree, y=y, group=assoc, colour=values)) +
    geom_line(alpha=1) +  
    scale_color_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = c(0,0)) + 
    theme_void() +
    theme(legend.position = "bottom")

triozid_plant_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
triozid_plant_tanglegram 
```

Next- move everything to this style fof tanglegram

## Community assembly

```{r}

# Weighted mpd
w_mpd <- picante::ses.mpd(comm, phydist, null.model = "taxa.labels", abundance.weighted = TRUE, runs = randoms)

#Unweighted mpd
u_mpd <- picante::ses.mpd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = randoms) 

# Weighted mntd
w_mntd <- picante::ses.mntd(comm, phydist, null.model = "taxa.labels", abundance.weighted = TRUE, runs = randoms) 

# Unweighted NRI
u_mntd <- picante::ses.mntd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = randoms) 

```

## Map of colleciton locations

```{r Collection map}
#Make spatial distance matrix
library(sp)
 # Spatial distance
  envData <- read.csv("sample_data/Sample_info.csv", header=TRUE)  %>%
    dplyr::filter(!duplicated(SampleID)) %>%
    set_rownames(.$SampleID) %>%
    separate(Collection, into= c("Lat", "Long"), sep=" ", remove = FALSE) %>%
    separate(Lat, paste("lat",c("d","m","s"), sep="_") ) %>%
    separate(Long, paste("long",c("d","m","s"), sep="_" ) ) %>%
    mutate_at(vars(-SampleID, -psyllid_spp), funs(as.numeric)) %>%
    mutate(lat=-(lat_d + lat_m/60 + lat_s/60^2),
              long=long_d + long_m/60 + long_s/60^2) %>%
    dplyr::select(SampleID, psyllid_spp, lat, long) %>%
    column_to_rownames("SampleID") %>%
    drop_na()

#Plot on map to confirm points
library("maptools")
data(wrld_simpl)
wrld_simpl@data$id <- wrld_simpl@data$NAME
wrld <- fortify(wrld_simpl, region="id")

gg.worldmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-90,90), xlim=c(-180,180))  

gg.anzmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-50,-30), xlim=c(130,180))+
  theme(legend.position = "bottom")  


gg.nzmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-50,-30), xlim=c(160,180)) 

```


```{r session-info}
# Display current R session information
sessionInfo()
```
