---
title: "Psyllid microbiome Statistics"
author: "Alexander Piper"
date: "14/08/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Load packages 

```{r install & Load packages} 
#Set required packages
.cran_packages <- c("ggplot2", "gridExtra", "tidyverse", "scales", "stringdist", "patchwork", "vegan", "ggpubr", "seqinr", "viridis", "ape", "sp", "data.table", "RColorBrewer", "ggtree")
.bioc_packages <- c("dada2", "phyloseq", "DECIPHER","Biostrings","ShortRead","psadd","microbiome","philr")

sapply(c(.cran_packages,.bioc_packages), require, character.only = TRUE)

#Load helper functions 
source('R/helper_functions.R')

options(stringsAsFactors = FALSE)

```



### Read in phyloseq and Summarise taxonomic assignment

```{r sum taxa}
ps <- readRDS("output/rds/ps_rdp.rds")

#Fraction of reads assigned to each taxonomic rank
sum_reads <- speedyseq::psmelt(ps) %>%
  gather("Rank","Name",rank_names(ps)) %>%
  group_by(Rank) %>% 
  mutate(Name = replace(Name, str_detect(Name, "__"),NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  summarize(Reads_classified = sum(Abundance * !is.na(Name))) %>%
  mutate(Frac_reads = Reads_classified / sum(sample_sums(ps))) %>%
  mutate(Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank)

#Fraction of ASV's assigned to each taxonomic rank
sum_otu <- tax_table(ps) %>%
  as("matrix") %>%
  as_tibble(rownames="OTU") %>%
  gather("Rank","Name",rank_names(ps)) %>%
  group_by(Rank) %>%
  mutate(Name = replace(Name, str_detect(Name, "__"),NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  summarize(OTUs_classified = sum(!is.na(Name))) %>%
  mutate(Frac_OTUs = OTUs_classified / ntaxa(ps)) %>%
  mutate(Rank = factor(Rank, rank_names(ps))) %>%
  arrange(Rank)

print(sum_reads)
print(sum_otu)
```


## Detect and remove outlier Samples

Detecting and potentially removing samples outliers (those samples with underlying data that do not conform to experimental or biological expectations) can be useful for minimizing technical variance. This can be caused by a number of reasons, including low-reads assigned to that sample. In this case we remove all samples below 1000 reads, as these include all samples contributing to lower than usual ASV counts.

```{r sample-removal-identification}
## Remove mocks
rm_mocks <- c("mockA_S51", "MockEven_S193", "Mock_S192", "PCRctrl_S191", "MockStaggered_S194", "PCRctrl_S191", "91_S167")
nsamples(ps)
ps0 <- ps %>% subset_samples(!sample_names(ps) %in% rm_mocks) %>% #Remove mocks
       
        filter_taxa(function(x) mean(x) > 0, TRUE) #Drop missing taxa from table 
nsamples(ps0)

# Format a data table to combine sample summary data with sample variable data
ss.df <- merge(as.data.frame(sample_data(ps0)), data.frame("ASV" = sample_sums(ps0)), by ="row.names")

# Plot Count of ASVs and read s
threshold = 1000

gg.asv.boxplot <- merge(as.data.frame(sample_data(ps0)), data.frame("Reads" = sample_sums(ps0)), by ="row.names") %>%
  ggplot( aes(x= Sample.Name, y = Reads, color = Sample.Name, shape=replicated)) + 
  geom_boxplot(outlier.colour="RED", position = position_dodge(width = 0.8)) +
  geom_jitter(size = 2, alpha = 0.6) +
  scale_y_log10() +
  geom_hline(yintercept = threshold, lty = 2) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle=90, vjust=1))

print(gg.asv.boxplot)
## Plot richness

gg.rich <- plot_richness(ps0, measures="Observed")
print(gg.rich)

#Remove all samples under the minimum read threshold 
ps1 <- prune_samples(sample_sums(ps0)>=threshold, ps0) # 
ps1 <- filter_taxa(ps1, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
print(paste(nsamples(ps) - nsamples(ps1), " Samples and ", ntaxa(ps) - ntaxa(ps1), " taxa Dropped"))

```


## Process Replicates

### Compare reproducibilty of replicates

Rarefaction curves are useful	to	assess	sensitivity	of	sample	size	to	observed	alpha-diversity estimates.
```{r replicate reproducibility}
#Subset to replicated samples
ps.reps <- subset_samples(ps1, replicated == TRUE)

#Plot read differences between replicates
gg.reps <- plot_bar(ps.reps, x="Sample", y="Abundance", fill="Phylum") +
  facet_grid(~Sample.Name, drop=TRUE, scales="free_x") +
  geom_hline(yintercept=1000)

#Plot rarefaction curve
rarecurve(otu_table(ps.reps), step=50, cex=0.5)

#Rarefy replicates to same read depth
nspec <- specnumber(otu_table(ps.reps)) # observed number of species
raremax <- min(rowSums(otu_table(ps.reps)))
rare <- rrarefy(otu_table(ps.reps),raremax)

#PCA of rarefied samples
raredist <- vegan::vegdist(rare, method="euclidean")

r.pcx <- prcomp(raredist)

pc_samp <- data.frame(SampleID = rownames(r.pcx$x), r.pcx$x[, 1:2])%>%
  left_join(samdf, by="SampleID")
pc_otu <- data.frame(OTU = rownames(r.pcx$rotation), r.pcx$rotation[, 1:2])

# calculate percent variance explained for the axis labels
pc1 <- round(r.pcx$sdev[1]^2/sum(r.pcx$sdev^2),2)
pc2 <- round(r.pcx$sdev[2]^2/sum(r.pcx$sdev^2),2)
pc_xlab <- paste("PC1: ", pc1, sep="")
pc_ylab <- paste("PC2: ", pc2, sep="")

library(ggplot2)
gg.reps <- ggplot(data=pc_samp, aes(x=PC1, y=PC2, colour=SampleID, label=SampleID)) + 
  geom_point(alpha=0.5, size=3) +
  geom_text() +
  #geom_point(data=pc_otu,aes(PC1, PC2)) +
  theme_bw() + 
  geom_hline(yintercept = 0, linetype=2) +  
  geom_vline(xintercept = 0, linetype=2) +
  xlab(pc_xlab) + 
  ylab(pc_ylab) +
  theme(legend.position = "none")+
  coord_fixed(ratio=pc2/pc1) # Scale plot by variance explained

print(gg.reps)
```

### Merge technical replicates

With only 16 samples replicated, and only on the first 2 sequencing runs i dont think there is a way to explicitly take into account replicate variability, therefore all replicates were merged

```{r replicates}
# Merge replicates
ps.merged <- ps1 %>%
    merge_samples(group = "Sample.Name")

#This loses the sample metadata - Need to add it agian
samdf <- read.csv("sample_data/Sample_info.csv", header=TRUE) %>%
  filter(!duplicated(Sample.Name)) %>%
  set_rownames(.$Sample.Name) %>%
  dplyr::select(c("SampleID", "Sample.Name", "seqrun", "psyllid_spp", "psyllid_genus", "psyllid_family", "hostplant_spp","Collection","Collection.Date"))

sample_data(ps.merged) <- samdf
ps.merged <- filter_taxa(ps.merged, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
```

## Prevalence assesment

Identification of taxa that are poorly represented in an unsupervised manner can identify taxa that will have little to no effect on downstream analysis. Sufficient removal of these "low prevalance" features can enhance many analysis by focusing statistical testing on taxa common throughout the data. However, for our dataset the problem with prevalence filtering for our dataset is that we dont have many replicates of each psyllid species, and therefore there are some high abundance but low prevalence ASV's we dont want to lose.

If we are agglomerating anyway, the prevalence filter doesnt have any impact really


```{r prevalence-assessment}
# Calculate taxon prevalence across the data set
prevdf <- apply(X = otu_table(ps.merged), MARGIN = ifelse(taxa_are_rows(ps.merged), yes = 1, no = 2), FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf - change this to tidyverse code
prevdf <- data.frame(Prevalence = prevdf, TotalAbundance = taxa_sums(ps.merged), tax_table(ps.merged))

#Prevalence plot
gg.prev <- subset(prevdf, Phylum %in% get_taxa_unique(ps0, "Phylum")) %>%
  ggplot(aes(TotalAbundance, Prevalence / nsamples(ps.merged),color=Genus)) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) +
  theme(legend.position="none") +
  ggtitle("Phylum Prevalence in All Samples\nColored by Family")

gg.prev
```

## Taxon Filtering

The following R chunk removes taxa not-typically part of a bacterial microbiome analysis.

We also remove all taxa contained within the blank sample from other samples

```{r taxon-cleaning}
get_taxa_unique(ps.merged, "Root")
get_taxa_unique(ps.merged, "Class")

ps.merged # Check the number of taxa prior to removal
ps2 <- ps.merged %>%
  subset_taxa(
    Root == "Bacteria" & #This is probably the only required one
    Family  != "Mitochondria" &
    Class   != "Chloroplast" &
    Order   != "Chloroplast" &
    Phylum != "Cyanobacteria"
  )
ps2 # Confirm that the taxa were removed
get_taxa_unique(ps2, "Phylum")
get_taxa_unique(ps2, "Class")

```


# Alpha Diversity

Instead of using traditional plug-in diversity estiamtors, which only tell us about the diversity of our samples (observed richness), instead of the actual environment (actual richness), we will use the breakaway package which uses biological replicates to estimate the total diversity in the environment. Breakaway estimates the number of missing species based on the sequence depth and  number of rare taxa in the data, and accounts for experimental noise such as the correlation between observed richness and the sequencing depth.

Following the initial estimate of diversity and variance using the breakaway package, we use the betta model of Willis, A. and Bunge, J. (2015) which works like regression but takes into account uncertainty in diversirty estimates betta accounts for strains that are present in the environment but not observed in the samples due to incomplete sampling.. Some of the samples in the initial estimates had large standard errors, indicating that we cannot precisely estimate the total species richness in those sample. In the betta model these samples will be given less weight in the regression than samples with small standard errors on their richness estimates. This is the behaviour we want, because while that sample contains some information, the standard error is telling us that the richness estimate may not be as reliable as other samples' estimates, and we should downweight the importance of that observation relative to observations would smaller standard errors on their richness estimates.

May be interesting to model richness at different taxonomic ranks?

Psyllid_spp and hostplant_spp are multi-colinear, therefore they cant be included as either additive effects or interaction effects, as it causes the design-matrix to have less than full rank, which means it isn't invertible, and therefore least-squares-techniques fail

This is a univariate phylogenetic analysis, and could be done with the package phylosignal, or a phylogenetic linear mixed model

The effects of phylogenetic signal, defined as “a tendency for related species to resemble each other more than they resemble species drawn at random from the tree” (74), on univariate traits (e.g., microbial alpha diversity) have been examined in parallel with phylosymbiosis studies. Phylogenetic signal indices like Pagel’s λ (75), and Blomberg’s K (76) are based on a random Brownian model of trait evolution (77), These can be calculated with with the phylosignal package

```{r estimate richness}
library(breakaway)

#Model richness for all samples by psyllid spp

ba <- breakaway(ps2)

## Check the specification of the model
betta_pic(summary(ba)$estimate, summary(ba)$error)

ba_estimates <- summary(ba) %>%
      add_column("Sample.Name" = ps2 %>% otu_table %>% sample_names) %>%
      left_join(samdf, by = "Sample.Name")

## Plot the richness estimates from each sample
col <- colorRampPalette(brewer.pal(11, "Spectral"))(75)
gg.barichnes <- ggplot(ba_estimates, aes(x=psyllid_spp, y=estimate, fill=psyllid_spp)) + 
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(aes(colour=psyllid_spp), colour="black", pch=21) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, vjust=0, hjust=1),
        legend.position = "none") +
  scale_fill_manual(values=col) +
  scale_colour_manual(values=col) +
  ylab("Richness Estimate")


#Model richness with psyllid_spp as a fixed effect
bt_spp <- betta(chats = ba_estimates$estimate,
                      ses = ba_estimates$error,
                      X = model.matrix(~psyllid_spp, data = ba_estimates)) 

bt_spp$table

bt_sppdf <- as_tibble(bt_spp$table, rownames="psyllid_spp")%>% 
                mutate_at(2:4, as.numeric) %>%
              magrittr::set_colnames(c("Estimates", "SE", "P"))

## Model with psyllid_spp as a fixed effect and seqrun as a random effect
bt_sppfixed_hostrandom <- betta_random(chats = ba_estimates$estimate,
                                  ses = ba_estimates$error,
                                  X = model.matrix(~psyllid_spp, data = ba_estimates), # Set fixed effects
                                  groups=ba_estimates$seqrun) # Set random effects


bt_randomdf <- as_tibble(bt_sppfixed_hostrandom$table, rownames="psyllid_spp")%>% 
                mutate(psyllid_spp = str_replace(psyllid_spp, pattern="psyllid_spp", replacement="")) %>%
                magrittr::set_colnames(c("psyllid_spp", "Estimates", "SE", "P")) %>%
                mutate(Estimates = case_when(
                  psyllid_spp == "(Intercept)" ~ Estimates,
                  !psyllid_spp == "(Intercept)" ~ Estimates + 
                                      filter(., str_detect(psyllid_spp, "Intercept") ) %>% # Rescale by intercept
                                      pull(Estimates)
                  
                )) %>%
                mutate(psyllid_spp = str_replace_all(psyllid_spp, pattern="\\(Intercept\\)", replacement =
                                                   unique(ba_estimates$psyllid_spp[which(!ba_estimates$psyllid_spp %in% .$psyllid_spp)])
                                                 )) %>%
                mutate(significant = case_when(
                  P < 0.05 ~ TRUE,
                  P > 0.05 ~ FALSE
                ))
#Our intercept term is one that is left out
intercept <- unique(ba_estimates$psyllid_spp[which(!ba_estimates$psyllid_spp %in% bt_randomdf$psyllid_spp)])

## Plotting - arrange by tree
psyllid_tree <- read.tree(text=readLines("sample_data/psyllid_beast_tree.nwk"))

# Match names with alpha diversity
psyllid_tree$tip.label <- psyllid_tree$tip.label %>%
  str_replace_all(pattern="\\.", replacement=" ") %>%
  str_replace_all(pattern="sp$", replacement="sp. ") %>%
  str_replace_all(pattern="Acizzia hakae", replacement="Acizzia hakeae") %>%
  str_replace_all(pattern="POLLENISLAND", replacement="POLLEN ISLAND") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiae$", replacement="Ctenarytaina fuchsia A") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeB", replacement="Ctenarytaina fuchsia B") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeC", replacement="Ctenarytaina fuchsia C") %>%
  str_replace_all(pattern="Ctenarytaina clavata", replacement="Ctenarytaina clavata sp. ") %>%
  str_replace_all(pattern="Ctenarytaina clavata sp. $", replacement="Ctenarytaina clavata sp. A") %>%
  str_replace_all(pattern="Ctenarytaina sp$", replacement="Ctenarytaina sp. ") %>%
  str_replace_all(pattern="Ctenarytaina unknown", replacement="Ctenarytaina insularis") %>%  
  str_replace_all(pattern="Psylla apicalis A", replacement="Psylla frodobagginsi") %>%
  str_replace_all(pattern="Psylla apicalis B", replacement="Psylla apicalis") %>%
  str_replace_all(pattern="carmichaeliae", replacement="carmichaeliae ") %>%
  str_replace_all(pattern="Psylla apicalis", replacement="Psylla apicalis ") %>% # Should these be A,B or?
  str_replace_all(pattern="Trioza sp", replacement="Trioza sp. ") %>%
  str_replace_all(pattern="Trioza acutaB", replacement="Trioza Acuta B") %>%
  str_replace_all(pattern="BRENDAMAY", replacement="BRENDA MAY") %>%
  str_replace_all(pattern="PRICES", replacement="PRICE'S VALLEY") %>%
  trimws(which="right")


psyllid_tree$tip.label[!psyllid_tree$tip.label %in% bt_randomdf$psyllid_spp]

pruned.tree <- drop.tip(psyllid_tree, psyllid_tree$tip.label[!psyllid_tree$tip.label %in% bt_randomdf$psyllid_spp] )
psyllid_tree$tip.label[!psyllid_tree$tip.label %in%  pruned.tree$tip.label]

# Plot tree + Richness estimates
library(ggtree)
p <- ggtree(pruned.tree) + geom_tiplab() + geom_nodelab(geom='label') +
    scale_x_continuous(expand=c(0, 0.1))

gg.alphatree <- facet_plot(p, panel='Species Richness', data = bt_randomdf %>% mutate(id = psyllid_spp) , geom=ggstance::geom_pointrangeh,
                 mapping=aes(x=Estimates, xmin=Estimates-SE, xmax=Estimates+SE, colour=significant), stat="identity")

test <- facet_plot(gg.alphatree, panel='Alpha Diversity',
           data=bt_randomdf, geom=geom_text, 
           mapping=aes(x=Estimates, label=round(Estimates)))


p4 <- facet_plot(gg.alphatree, panel="PD", data=pd.plot %>% mutate(id = psyllid_spp) , ggstance::geom_boxploth, 
            mapping = aes(x=PD, group=id))

p5 <- facet_plot(p4, panel="SR", data=pd.plot %>% mutate(id = psyllid_spp) , ggstance::geom_boxploth, 
            mapping = aes(x=SR, group=id))


# Could probably highlight the few clades with a high local morans I in Red for the plot

# Calculate Faith's PD-index
pdtable <- picante::pd(as(phyloseq::otu_table(ps2), "matrix"), phyloseq::phy_tree(ps2), include.root = F)

# Calculate Faith's PD-index with SES
test <- picante::ses.pd(as(phyloseq::otu_table(ps2), "matrix"), phyloseq::phy_tree(ps2), include.root = F, runs=10)

pd.plot <- test %>%
      rownames_to_column("Sample.Name") %>%
      left_join(samdf, by = "Sample.Name") %>%
  select(psyllid_spp, pd.obs, pd.obs.z, pd.obs.p, Sample.Name)

#Model richness with psyllid_spp as a fixed effect
pd_spp <- betta(chats = pd.plot$estimate,
                      ses = ba_estimates$error,
                      X = model.matrix(~psyllid_spp, data = ba_estimates)) 


# Use this on the dotplot

pd.plot <- pdtable %>%
      rownames_to_column("Sample.Name") %>%
      left_join(samdf, by = "Sample.Name") %>%
  select(psyllid_spp, PD, SR, Sample.Name)
## Plot the PD estimates from each sample
col <- colorRampPalette(brewer.pal(11, "Spectral"))(75)
gg.barichnes <- ggplot(pd.plot, aes(x=psyllid_spp, y=PD, fill=psyllid_spp)) + 
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(aes(colour=psyllid_spp), colour="black", pch=21) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, vjust=0, hjust=1),
        legend.position = "none") +
  scale_fill_manual(values=col) +
  scale_colour_manual(values=col) +
  ylab("Richness Estimate")

# Plug in estimators


# Add Statistics
library(phylosignal)
library(adephylo)
#alpha diversity data
dat <- list()
dat$alpha <-  bt_randomdf %>%
  filter(psyllid_spp %in% pruned.tree$tip.label) %>%
  arrange(match(psyllid_spp, pruned.tree$tip.label)) %>%
  pull(Estimates)
dat$random <- rnorm(length(dat$alpha), sd = 10) # NEgative control
dat$bm <- rTraitCont(pruned.tree) #Brownian motion - Positive control
dat <- as.data.frame(dat)

# Make phylosignal object and measure signal

p4d <- phylobase::phylo4d(pruned.tree, dat)
phyloSignal(p4d = p4d, method = "all")

# Asses modelbehavior
phylosim <- phyloSim(tree = pruned.tree, method = "all", nsim = 100, reps = 99)
plot.phylosim(phylosim, what = "pval", stacked.methods = TRUE)
```


RESULTS: while there were significant differences between psyllid species (see betta results), there were no significant associations with phylogeny

$stat
             Cmean            I          K    K.star    Lambda
alpha  -0.01056585 -0.025740055 0.70501041 0.3313264 0.8755386
random  0.06885442  0.001435941 0.03960619 0.2290344 0.2717360
bm      0.42031946  0.145456264 0.14547611 0.9100803 1.0020607

$pvalue
       Cmean     I     K K.star    Lambda
alpha  0.445 0.634 0.017  0.208 0.0010000
random 0.166 0.125 0.472  0.539 0.1048509
bm     0.001 0.001 0.029  0.001 0.0010000


# Beta diversity


To find the Aitchison distance a centered log-ratio (clr) transformation was performed on ASV count data with the R
165 package CoDaSeq (30) before computing the Euclidean distance between samples with
the R package vegan (31). The Aitchison distance was then used in Permutational Multivariate Analyses of Variance (PERMANOVA, 999 permutations) with the vegan function adonis to test whether community composition was significantly different

Many analysis in community ecology and hypothesis testing benefit from data transformation. Many microbiome data sets do not fit to a normal distribution, but transforming them towards normality may enable more appropriate data for specific statistical tests. 

philr phylogenetic transform is based on balances (binary partitions) along an evolutionary tree (Silverman et al., 2017) that is a replacement for the familiar UniFrac distance metric. Distances determined by phylogenetic transforms have the advantage that the binary partitions chosen have a simple interpretation and the correlation structure of the data is fully accounted for. However, the disadvantage is that only the relationships between the chosen partitions can be examined.# Figure 1: Community composition analysis

## PhilR transformation

- might be worth doing these with both the CLR and philr transformed data

```{r Philr transfomration}
ps.coda <- tip_glom(ps2, h=0.3) # Work out what h=0.3 approximates

#Rename taxa
taxa_names(ps.coda) <- paste0("SV", seq(ntaxa(ps.coda)),"-",tax_table(ps.coda)[,6])

ps2.n0 <- transform_sample_counts(ps.coda, function(x) x+1)

phy_tree(ps2.n0) <- multi2di(phy_tree(ps2.n0))
phy_tree(ps2.n0) <- makeNodeLabel(phy_tree(ps2.n0), method="number", prefix='n')

#PhilR distance
otu.table <- otu_table(ps2.n0)
tree <- phy_tree(ps2.n0)
metadata <- as(sample_data(ps2.n0), "data.frame")
tax <- tax_table(ps2.n0)
ps2.philr <- philr(otu.table, tree,
                  part.weights='enorm.x.gm.counts',
                  ilr.weights='blw.sqrt')

gp.dist <- dist(ps2.philr, method="euclidean")


# Ordination
gp.pcoa <- ordinate(ps.coda, 'PCoA', distance=gp.dist)
gg.pca <- plot_ordination(ps.coda, gp.pcoa, color='psyllid_spp', label="psyllid_spp", type="Sample") + geom_point(size=4)

# Fviz biplot
library(factoextra)
f.pcx <- prcomp(ps2.philr)
pcx <- f.pcx 

#Name balances
pcnames <-  sapply(rownames(pcx$rotation), function(x) name.balance(tree, tax, x))
pcnames <- make.unique(pcnames, sep = ".")
rownames(pcx$rotation) <- pcnames

fviz_pca_biplot(pcx) 

```

### Create distance matrices

Note: Alpha-diversity could also be converted to a distance matrix by taking the Euclidean distance among all pairwise sample comparisons.

```{r distance matrices}
metadata <- as(sample_data(ps2), "data.frame")

# COI Genetic distance
phylo.dist <- read_csv("sample_data/genetic_distance.csv", col_names = FALSE) %>%
  column_to_rownames("SampleID") 

phylo.dist[upper.tri(phylo.dist)] <- 0
phylo.dist <- phylo.dist + t(phylo.dist)
phylo.dist[is.na(phylo.dist)] <- 0

phylo.dist <- phylo.dist %>%
  set_colnames(rownames(.) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(gp.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(gp.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(gp.dist)))) %>%
  as.dist()

phylo.dist#check whats missing
colnames(as.matrix(comm.dist))[which(!colnames(as.matrix(comm.dist)) %in% colnames(as.matrix(phylo.dist)))]


# Distance from the bayesian tree - try join this to the actual samples
pruned.tree
cophenetic <- cophenetic(pruned.tree)

tree.dist <- samdf %>%
  select(SampleID, psyllid_spp) %>%
  dplyr::rename(psyllid_spp1 = psyllid_spp) %>%
  left_join(cophenetic %>%
  as.data.frame() %>%
  rownames_to_column("psyllid_spp1") %>%
  tidyr::pivot_longer(-psyllid_spp1, names_to = "psyllid_spp2",
               values_to = "dist")
  ) %>%
  left_join(
    samdf %>%
  select(SampleID, psyllid_spp) %>%
  dplyr::rename(psyllid_spp2 = psyllid_spp) %>%
  dplyr::rename(SampleID2 = SampleID)
  ) %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% str_to_upper()) %>%
  mutate(SampleID2 = SampleID2 %>% str_replace(pattern="\\_S(.*)$",replacement="")%>% str_to_upper()) %>%
  select(-one_of("psyllid_spp1", "psyllid_spp2")) %>%
  #unique() %>%
  pivot_wider(names_from = SampleID2,
              values_from = dist) %>%
  filter(SampleID %in% rownames(as.matrix(gp.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(gp.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(gp.dist)))) %>%
  as.dist()

# Microbial community dist
comm.dist <- gp.dist %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  filter(SampleID %in% rownames(as.matrix(phylo.dist))) %>%
  dplyr::select(one_of(colnames(as.matrix(phylo.dist)))) %>%
  as.dist()


# Hostplant distance
plant.dist <- read_csv("sample_data/plant_distance.csv") %>%
  dplyr::select(-one_of("X1", "Host plants")) %>%
  column_to_rownames("SampleID") %>%
  set_colnames(rownames(.) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(comm.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(comm.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(comm.dist)))) %>%
  as.dist()

# Spatial distance
envData <- read.csv("sample_data/Sample_info.csv", header=TRUE)  %>%
  dplyr::filter(!duplicated(SampleID)) %>%
  set_rownames(.$SampleID) %>%
  separate(Collection, into= c("Lat", "Long"), sep=" ", remove = FALSE) %>%
  separate(Lat, paste("lat",c("d","m","s"), sep="_") ) %>%
  separate(Long, paste("long",c("d","m","s"), sep="_" ) ) %>%
  mutate_at(vars(-SampleID, -psyllid_spp), funs(as.numeric)) %>%
  mutate(lat=-(lat_d + lat_m/60 + lat_s/60^2),
            long=long_d + long_m/60 + long_s/60^2) %>%
  select(SampleID,psyllid_spp, lat, long) %>%
  column_to_rownames("SampleID") %>%
  drop_na()


#Make spatial distance matrix
library(sp)
spat.dist <- spDists(as.matrix(envData %>% select(-psyllid_spp)), longlat=TRUE) %>%
  as.data.frame() %>%
  set_rownames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  set_colnames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(comm.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(comm.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(comm.dist)))) %>%
  as.dist()


#Plot on map to confirm points
library("maptools")
data(wrld_simpl)
wrld_simpl@data$id <- wrld_simpl@data$NAME
wrld <- fortify(wrld_simpl, region="id")

gg.worldmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-90,90), xlim=c(-180,180))  

gg.anzmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-50,-30), xlim=c(130,180))+
  theme(legend.position = "bottom")  


gg.nzmap <- ggplot(envData) +
  geom_map(data=wrld, map=wrld, aes(map_id=id, x=long, y=lat), fill="lightgrey", color="#7f7f7f", size=0.25) +
  geom_point(aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3) + 
  theme_bw() +
  coord_fixed(ylim = c(-50,-30), xlim=c(160,180)) 

```


## Relationship over agglomeration

```{r agglomeration}
## CLR
f <- t(abundances(ps.coda))
#f.0 <- cmultRepl(f, method="CZM")
f.0 <- f
f.clr <- codaSeq.clr(f.0)
#perb.all <- perb(f.0, ivar="iqlr")
clr.dist <- dist(f.clr, method="euclidean")

##


set.seed(666)
glomseq <- seq(0, 0.8,0.1)

adlist <- vector("list", length=length(glomseq))
mantlist <- vector("list", length=length(glomseq))
treelist <- vector("list", length=length(glomseq))

for (i in 1:length(glomseq)){
  
ps.glom <- tip_glom(ps2, h=glomseq[i])

treelist[i] <- phy_tree(psglom)

#Rename taxa
taxa_names(ps.glom) <- paste0("SV", seq(ntaxa(ps.glom)),"-",tax_table(ps.glom)[,6])

ps2.philr <- transform_sample_counts(ps.glom, function(x) x+1)

phy_tree(ps2.philr) <- multi2di(phy_tree(ps2.philr))
phy_tree(ps2.philr) <- makeNodeLabel(phy_tree(ps2.philr), method="number", prefix='n')

#Transpose
otu.table <- otu_table(ps2.philr)
tree <- phy_tree(ps2.philr)
metadata <- as(sample_data(ps2.philr), "data.frame")
tax <- tax_table(ps2.philr)
ps.philr <- philr(otu.table, tree,
                  part.weights='enorm.x.gm.counts',
                  ilr.weights='blw.sqrt')

distances <-  c("ps2.philr", "clr.ds"
for (d in distances)

gp.dist <- dist(ps2.philr, method="euclidean")


# ADONIS TESTS
metadata <- as(sample_data(ps2.philr), "data.frame")


ad_psyllid <- adonis(gp.dist ~ psyllid_spp, method = "euclidean",
                   data = metadata)$aov.tab %>%
  as_tibble() %>%
  mutate(call = "psyllid_spp")


ad_hostplant <- adonis(gp.dist ~ hostplant_spp, method = "euclidean",
                     data = metadata)$aov.tab %>%
  as_tibble() %>%
  mutate(call = "hostplant_spp")

#Interactions
ad_psyllid_hostplant <- adonis(gp.dist ~ psyllid_spp*hostplant_spp, method = "euclidean",
                        data = metadata)$aov.tab %>%
  as_tibble() %>%
  mutate(call = "psyllid_spp*hostplant_spp")

ad_hostplant_psyllid <- adonis(gp.dist ~ hostplant_spp*psyllid_spp,
                        data = metadata)$aov.tab %>%
  as_tibble() %>%
  mutate(call = "hostplant_spp*psyllid_spp")


adlist[[i]] <- bind_rows(ad_psyllid, ad_hostplant, ad_psyllid_hostplant, ad_hostplant_psyllid) %>%
  mutate(glom = glomseq[i])


## Mantel & Procrustes

## Make distance matrices
# COI Genetic distance
phylo.dist <- read_csv("sample_data/genetic_distance.csv", col_names = TRUE) %>%
  column_to_rownames("SampleID") 

phylo.dist[upper.tri(phylo.dist)] <- 0
phylo.dist <- phylo.dist + t(phylo.dist)
phylo.dist[is.na(phylo.dist)] <- 0

phylo.dist <- phylo.dist %>%
  set_colnames(rownames(.) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(gp.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(gp.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(gp.dist)))) %>%
  as.dist()

# Microbial community dist
comm.dist <- gp.dist %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  filter(SampleID %in% rownames(as.matrix(phylo.dist))) %>%
  dplyr::select(one_of(colnames(as.matrix(phylo.dist)))) %>%
  as.dist()

# Hostplant distance
plant.dist <- read_csv("sample_data/plant_distance.csv") %>%
  dplyr::select(-one_of("X1", "Host plants")) %>%
  column_to_rownames("SampleID") %>%
  set_colnames(rownames(.) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(comm.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(comm.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(comm.dist)))) %>%
  as.dist()

# Spatial distance
envData <- read.csv("sample_data/Sample_info.csv", header=TRUE)  %>%
  dplyr::filter(!duplicated(SampleID)) %>%
  set_rownames(.$SampleID) %>%
  separate(Collection, into= c("Lat", "Long"), sep=" ", remove = FALSE) %>%
  separate(Lat, paste("lat",c("d","m","s"), sep="_") ) %>%
  separate(Long, paste("long",c("d","m","s"), sep="_" ) ) %>%
  mutate_at(vars(-SampleID, -psyllid_spp), funs(as.numeric)) %>%
  mutate(lat=-(lat_d + lat_m/60 + lat_s/60^2),
            long=long_d + long_m/60 + long_s/60^2) %>%
  select(SampleID,psyllid_spp, lat, long) %>%
  column_to_rownames("SampleID") %>%
  drop_na()

#Make spatial distance matrix
spat.dist <- spDists(as.matrix(envData %>% select(-psyllid_spp)), longlat=TRUE) %>%
  as.data.frame() %>%
  set_rownames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  set_colnames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  filter(SampleID %in% rownames(as.matrix(comm.dist))) %>%
  arrange(match(SampleID, colnames(as.matrix(comm.dist)))) %>%
  column_to_rownames("SampleID") %>%
  dplyr::select(one_of(colnames(as.matrix(comm.dist)))) %>%
  as.dist()

## Mantel tests

#Microbial community ~ Phylogenetic distance
man_phylo <- mantel(comm.dist, phylo.dist)
man_phylo <- data.frame(
  call="M Microbe~Phylo",
  test="Mantel",
  stat=man_phylo$statistic,
  signif=man_phylo$signif)

#Microbial Community ~ Plant Distance
man_plant <- mantel(comm.dist, plant.dist)
man_plant <- data.frame(
  call="M Microbe~Plant",
  test="Mantel",
  stat=man_plant$statistic,
  signif=man_plant$signif)

#Microbial community ~ Geographic distance
man_geo <- mantel(comm.dist, spat.dist)
man_geo <- data.frame(
  call="M Microbe~Geography",
  test="Mantel",
  stat=man_geo$statistic,
  signif=man_geo$signif)

## Partial mantel tests

#Partial mantel taking into account plant
pman_phylo_plant <- mantel.partial(comm.dist, phylo.dist, plant.dist)
pman_phylo_plant <- data.frame(
  call="PM Microbe~Phylo~Plant",
  test="Partial Mantel",
  stat=pman_phylo_plant$statistic,
  signif=pman_phylo_plant$signif)

#Partial mantel taking into account geography
pman_phylo_geo <- mantel.partial(comm.dist, phylo.dist, spat.dist)
pman_phylo_geo <- data.frame(
  call="PM Microbe~Phylo~Geography",
  test="Partial Mantel",
  stat=pman_phylo_geo$statistic,
  signif=pman_phylo_geo$signif)

#Partial mantel taking into account phylogeny
pman_plant_phylo <- mantel.partial(comm.dist, plant.dist, phylo.dist)
pman_plant_phylo <- data.frame(
  call="PM Microbe~Plant~Phylo",
  test="Partial Mantel",
  stat=pman_plant_phylo$statistic,
  signif=pman_plant_phylo$signif)

#Partial mantel taking into account geography
pman_plant_geo <- mantel.partial(comm.dist, plant.dist, spat.dist)
pman_plant_geo <- data.frame(
  call="PM Microbe~Plant~Geography",
  test="Partial Mantel",
  stat=pman_plant_geo$statistic,
  signif=pman_plant_geo$signif)

#Partial mantel taking into host phylogeny
pman_geo_phylo <- mantel.partial(comm.dist, spat.dist, phylo.dist)
pman_geo_phylo <- data.frame(
  call="PM Microbe~Geography~Phylo",
  test="Partial Mantel",
  stat=pman_geo_phylo$statistic,
  signif=pman_geo_phylo$signif)

#Partial mantel taking into account plant phylogeny
pman_geo_plant <- mantel.partial(comm.dist, spat.dist, plant.dist)
pman_geo_plant <- data.frame(
  call="PM Microbe~Geography~Plant",
  test="Partial Mantel",
  stat=pman_geo_plant$statistic,
  signif=pman_geo_plant$signif)

##Procrustes tests

phylo_proc <- protest(monoMDS(comm.dist), monoMDS(phylo.dist)) # Do i need the MDS scaling?
phylo_proc <- data.frame(
  call="Procrustes Microbe~Phylo",
  test="Procrustes",
  stat=phylo_proc$ss,
  scale=phylo_proc$scale,
  signif=phylo_proc$signif)

plant_proc <- protest(monoMDS(comm.dist), monoMDS(plant.dist))
plant_proc <- data.frame(
  call="Procrustes Microbe~Plant",
  test="Procrustes",
  stat=plant_proc$ss,
  scale=plant_proc$scale,
  signif=plant_proc$signif)


#Procrustes
geo_proc <- protest(monoMDS(comm.dist), monoMDS(spat.dist))
geo_proc <- data.frame(
  call="Procrustes Microbe~Geography",
  test="Procrustes",
  stat=geo_proc$ss,
  scale=geo_proc$scale,
  signif=geo_proc$signif)


mantlist[[i]] <- bind_rows(man_phylo, man_geo, man_plant,
                  pman_phylo_geo, pman_phylo_plant,
                  pman_plant_geo, pman_plant_phylo,
                  pman_geo_phylo, pman_geo_plant,
                  phylo_proc, plant_proc, geo_proc) %>%
  mutate(glom = glomseq[i])

}


adonis_stats <- bind_rows(adlist)

gg.adonis <- ggplot(adonis_stats, aes(x=glom, y=R2, fill=call))

matrix_stats <- bind_rows(mantlist) %>%
  mutate(signif_test = case_when(
    signif < 0.01 ~ TRUE,
    signif > 0.01 ~ FALSE
  ))


mantel <- matrix_stats %>% 
  filter(test %in% c("Mantel", "Partial Mantel"))

gg.mantel <- ggplot(mantel, aes(x=call, y=stat, fill=call)) + 
  geom_bar(stat="identity") + 
  geom_point(data=mantel %>% filter(signif_test==TRUE), aes(x = call, y=stat + 0.005), shape=8, colour="black") +
  scale_fill_brewer(palette="Spectral") + 
  facet_grid(~glom) +
  ylab("Mantel test statistic") +
  theme_classic() + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

# Trees
treelist


library(ggtree)
trees <- treelist
class(trees) <- "multiPhylo"
ggtree(trees) + facet_wrap(~.id, scale="free") + geom_tiplab()

```


This suggests that species significantly cluster, and so do hostplants. However when accounting for species, hostplants are not significant. Therefore it should be justified to redo the barchart merging by species 


# Test for phylosymbiosis
Use mantel, Procrustean superimposition 


### Matrix correlation
Also see paco - https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12736

```{r matrix correlation}

#Microbial community ~ Phylogenetic distance
#Mantel test
plot(comm.dist~phylo.dist)
mantel(comm.dist, phylo.dist)
plot(mantel.correlog(comm.dist, phylo.dist), alpha=0.05)

#Partial mantel taking into account plant
mantel.partial(comm.dist, phylo.dist, plant.dist)

#Partial mantel taking into account geography
mantel.partial(comm.dist, phylo.dist, spat.dist)

#Procrustes
phylo_proc <- protest(monoMDS(comm.dist), monoMDS(phylo.dist)) # Do i need the MDS scaling?

summary(phylo_proc)
plot(phylo_proc)
plot(phylo_proc, kind=2)
residuals(phylo_proc)


#Microbial Community ~ Plant Distance
#Mantel test 
plot(comm.dist~plant.dist)
mantel(comm.dist, plant.dist)
plot(mantel.correlog(comm.dist, plant.dist), alpha=0.05)

#Partial mantel taking into account phylogeny
mantel.partial(comm.dist, plant.dist, phylo.dist)

#Partial mantel taking into account geography
mantel.partial(comm.dist, plant.dist, spat.dist)

#Procrustes
plant_proc <- protest(monoMDS(comm.dist), monoMDS(plant.dist))
plant_proc <- protest(comm.dist, plant.dist) # this gives very different results!

summary(plant_proc)
plot(plant_proc)
plot(plant_proc, kind=2)
residuals(plant_proc)

#Microbial community ~ Geographic distance
#Mantel test
plot(comm.dist~spat.dist)
plot(comm.dist~c(spat.dist+1), log="x")
mantel(comm.dist, spat.dist)
plot(mantel.correlog(comm.dist, spat.dist), alpha=0.05)

#Partial mantel taking into host phylogeny
mantel.partial(comm.dist, spat.dist, phylo.dist)

#Partial mantel taking into account plant phylogeny
mantel.partial(comm.dist, spat.dist, plant.dist)

#Procrustes
geo_proc <- protest(monoMDS(comm.dist), monoMDS(spat.dist))

summary(geo_proc)
plot(geo_proc)
plot(geo_proc, kind=2)
residuals(geo_proc) #Work out how to interpret these
```


### Tree topology comparison
robinson foulds, matching cluster

But also see how this paper has done it with PACO https://www.nature.com/articles/s41467-019-10191-3.pdf Using random subsampling? Then pull out the most associated also file:///C:/Users/ap0y/Dropbox/workprojects/Papers/10.1002_ecy.1955.pdf
```{r}

```


# Figures 

## Species Psyllid spp, beta diversity, alpha diversity  

Dendrogram and bar plots

Merge species, arrange microbiome by phylogeny, try the mantel tests on the merged data as well (but the seperate is probs more statistically better)

```{r merge spp}
# Merge species
ps.sppmerged <- ps2 %>%
    merge_samples(group = "psyllid_spp")

#This loses the sample metadata - Need to add it agian
samdf <- read.csv("sample_data/Sample_info.csv", header=TRUE) %>%
  filter(!duplicated(psyllid_spp)) %>%
  set_rownames(.$psyllid_spp) %>%
  dplyr::select(c("SampleID", "Sample.Name", "seqrun", "psyllid_spp", "psyllid_genus", "psyllid_family", "hostplant_spp","Collection","Collection.Date"))

sample_data(ps.sppmerged) <- samdf
ps4 <- filter_taxa(ps.sppmerged, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table

#Rename taxa
taxa_names(ps4) <- paste0("SV", seq(ntaxa(ps4)),"-",tax_table(ps4)[,6])

ps4_bar <- #subset_taxa(ps3, !Family == "Enterobacteriaceae") %>%
            ps4 %>%
  speedyseq::tax_glom(taxrank = "Order") %>%           # agglomerate at Order level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  speedyseq::psmelt() %>%
  mutate(plotlabel = Phylum) %>%
  mutate(plotlabel = case_when(
    Abundance >= 0.01 & Phylum=="Proteobacteria" ~ paste0(Phylum, " - ", Order), # Change this to whatever taxrank we want
    Abundance >= 0.01 & !Phylum=="Proteobacteria"~ Phylum ,
    Abundance < 0.01 ~ "NA"
    )) %>%
  dplyr::na_if("NA") %>%
  select(Sample, plotlabel, Phylum, Order, Abundance)

library(ggtree)
p <- ggtree(pruned.tree) + geom_tiplab(align=TRUE) + geom_nodelab(geom='label') +
    scale_x_continuous(expand=c(0, 0.1))


Fig1 <- facet_plot(p, panel='Relative Abundance', data = ps4_bar %>% mutate(id = Sample), geom=ggstance::geom_barh,
                 mapping=aes(x = Abundance, fill = plotlabel), stat="identity")  + 
  scale_fill_manual(values=colorRampPalette(brewer.pal(11, "Paired"))(length(unique(ps4_bar$plotlabel))-1), na.value="grey")


Fig1 <- facet_plot(Fig1, panel='Species Richness', data = bt_randomdf %>% mutate(id = psyllid_spp),
                          geom=ggstance::geom_pointrangeh,
                           mapping=aes(x=Estimates, xmin=Estimates-SE, xmax=Estimates+SE),
                           stat="identity", size=0.5, colour="#696969") +
  #scale_colour_manual(values=c("#808080","#9e0142"),na.value="#808080") + 
  theme_classic2()   +
  theme(legend.position = "bottom",
  panel.grid.major.x = element_line(colour="grey92", size=0.5, linetype="dashed"),
  strip.background = element_rect(fill = "grey92", 
                colour = "black", size = 1)
  )
        
# Change relative grid sizes
library(grid)
library(gtable)

gt = ggplot_gtable(ggplot_build(Fig1))
gt$layout$l[grep('panel-3', gt$layout$name)] # you want to find the column specific to panel-2

gt$widths[9] = 0.3*gt$widths[9] # reduce width of column 9
grid.draw(gt) # plot with grid draw

```


# Test for Cophylogeny

## Primary symbiont

Use paco https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12736

and Parafit https://www.rdocumentation.org/packages/ape/versions/5.3/topics/parafit

## Minimum abundance threshold

Maybe this should be done just for carsonella tree

These samples were sequenced using combinatorial indexes. Therefore this data will contain index switching. We therefore need to select an optimal filtering threshold to remove switched taxa, but not real taxa. We will optimise our filtering threshold by relying on assumption that all psyllid species should contain only one (or few very closely related ASVs) of Carsonella.

```{r thresholds}
#Flag top abundance carsonella
top_carson <-  ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  filter(Genus=="Candidatus_Carsonella") %>%
  group_by(Sample) %>%
  top_n(1, wt=Abundance) %>%
  mutate(top = TRUE)  %>%
  filter(Abundance > 0)
  
#Flag lower abundance carsonella
carson <-  ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  filter(Genus=="Candidatus_Carsonella") %>%
  left_join(top_carson) %>%
  mutate(top = case_when(
    is.na(top) ~ FALSE,
    !is.na(top) ~ top
  ))

## THis needs to be changed - things should be flagged as switched when less than 100* the top!, or combine both - ie 

carson <- carson %>%
  left_join(carson %>%
    filter(Abundance > 0) %>%
    dplyr::select(OTU, top, seqrun) %>%
    unique() %>%
    group_by(OTU, seqrun) %>%
    add_tally() %>%
    mutate(switched = case_when(
      n>1 & top== FALSE ~ TRUE,
      n>1 & top== TRUE ~ FALSE,
      n==1 ~ FALSE,
    )) %>%
  dplyr::select(-n)
  )

## Plot true and false carsonellas by run - this shows may need to filter seperately by run, as seqrun2 has much higher than others.
gg.switch <- carson %>% 
  group_by(seqrun, switched) %>%
  filter(!is.na(switched)) %>%
  filter(Abundance > 0) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  ggplot(aes(x=seqrun, y=freq, group=switched, fill=switched)) +
  geom_bar(stat="identity")


#to check if index switching could explain the multiple OTU's check if top=FALSE carsonella exist as top in another sample in the same run

thresholds <- seq(0,0.05,0.0001)
out <- vector("list", length=3)
for (r in 1:3) {
  df <- data.frame(thresh = thresholds, TP= thresholds, FP = thresholds, seqrun = r)
  for(i in 1:length(thresholds)){
    filt <- carson %>%
      filter(seqrun==r) %>%
      filter(Abundance > thresholds[i]) %>%
      group_by(top) %>%
      summarise(sum=n())
    df$FP[i] <- filt$sum[1]
    df$TP[i] <- filt$sum[2]
  }
  out[[r]] <- df
}

gg.filt <- out %>%
  bind_rows() %>%
  gather(key=type, value=n, -thresh, -seqrun) %>%
  ggplot(aes(x=thresh, y=n, fill=type)) + 
  geom_density(stat="identity", alpha=0.5) + 
  scale_x_continuous(breaks=seq(0,0.05,0.001)) +
  facet_grid(~seqrun) +
  theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  geom_vline(xintercept = 0.0035)+ 
  geom_vline(xintercept = 0.001)

print(gg.filt)

## Filter run 2 
run2_ps <- ps2 %>% 
  subset_samples(seqrun==2)

run2_pass <- run2_ps %>%
    transform_sample_counts(function (x) x/sum(x)) %>%  # Convert to proportions
    transform_sample_counts(function (x) (x > 0.001) * 1)

newotu1 <- otu_table(run2_ps) * otu_table(run2_pass)

run2_newps <- run2_ps
otu_table(run2_newps) <- otu_table(newotu1, taxa_are_rows = FALSE) 

#Filter run 1 and 3 
  
run13_ps <- ps2 %>%
    subset_samples(!seqrun == 2)
                   
run13_pass <- run13_ps %>%
    transform_sample_counts(function (x) x/sum(x)) %>%  # Convert to proportions
    transform_sample_counts(function (x) (x > 0.0005) * 1)

newotu2  <- otu_table(run13_ps) * otu_table(run13_pass)

run13_newps <- run13_ps
otu_table(run13_newps) <- otu_table(newotu2, taxa_are_rows = FALSE) 

# Create new phyloseq and drop missing taxa
ps3 <- merge_phyloseq(run2_newps, run13_newps) %>%
  filter_taxa(function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
ps3 <- prune_samples(sample_sums(ps3) >0 , ps3) # Drop empty samples

#Count number of overall taxa pre and post filtering
print(paste(ntaxa(ps2) - ntaxa(ps3), " taxa Dropped when using filtering threshold of: ", 0.0035, " for run 2 and ", 0.001, "for run 1 and 3"))

# Count number of carsonella OTU's per sample pre and post filtering
n_carson <- speedyseq::psmelt(ps2) %>% 
  filter(Genus == "Candidatus_Carsonella") %>%
  filter(Abundance > 0) %>%
  dplyr::select(Sample.Name, OTU) %>%
  group_by(Sample.Name) %>%
  add_tally() %>%
  dplyr::select(-OTU) %>%
  unique() %>%
  mutate(type = "pre") %>%
  bind_rows(.,speedyseq::psmelt(ps3) %>% 
  filter(Genus == "Candidatus_Carsonella") %>%
  filter(Abundance > 0) %>%
  dplyr::select(Sample.Name, OTU) %>%
  group_by(Sample.Name) %>%
  add_tally() %>%
  dplyr::select(-OTU) %>%
  unique() %>%
  mutate(type = "post"))

gg.ncarson <- ggplot(n_carson, aes(x=reorder(Sample.Name, -n), y=n)) + 
  geom_bar(stat="identity") +
  facet_grid(~type) + 
  xlab("Sample.Name") +
  ylab("Number of carsonella OTU's per sample") +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust=0))

print(gg.ncarson)

#check if any samples dont have carsonella
table(!n_carson$Sample.Name %in% speedyseq::psmelt(ps2)$Sample.Name)

## Get the name of taxa that dont have carsonella after filtering

```

```{r subsetting}
library(plotly)

#filter threshold to ge
#ps2.carson <- transform_sample_counts(ps2, fun = proportions, thresh=0.0045)

#ps2.carson <- subset_taxa(ps2.carson, Genus == "Candidatus_Carsonella")
#ps2.carson <- filter_taxa(ps2.carson, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
#ps2.carson <- prune_samples(sample_sums(ps2.carson)>0, ps2.carson) # Drop empty samples

library(RColorBrewer)
col <- colorRampPalette(brewer.pal(11, "Spectral"))(74)


ps2.carson <- subset_taxa(ps2, Species == "Candidatus Carsonella")
gg.carson <- plot_tree(ps2.carson, ladderize="left",  color="psyllid_spp") +   theme(legend.position = "none") + #
    scale_colour_manual(values=col) + scale_fill_manual(values=col)
  ggtitle("Carsonella 16S gene tree")
#gg.ncarson
  
## Aquabacterium
gg.tree <- plot_tree(subset_taxa(ps2, Species == "Sulfuritalea"), ladderize="left",  color="psyllid_spp") +   theme(legend.position = "none") + #
    scale_colour_manual(values=col) + scale_fill_manual(values=col)
  ggtitle("Carsonella 16S gene tree")
#gg.ncarson

#Write out fasta

carson_out <- speedyseq::psmelt(ps3.carson) %>%
  filter(Abundance > 0) %>%
  select(OTU, Sample.Name, psyllid_spp, Species) %>%
  group_by(OTU, Species) %>%
  summarize(samples = paste(Sample.Name, collapse=";"),
            psyllid_spp = paste(unique(psyllid_spp), collapse=";")
            ) %>%
  rownames_to_column() %>%
  unite(seqname, c("Species","rowname","psyllid_spp","samples"), sep="|")

carson_fasta <- DNAStringSet(carson_out$OTU)
names(carson_fasta) <- carson_out$seqname

writeXStringSet(carson_fasta, "carsonella_seqs.fa")


## S-Symbionts
ps3.carson <- subset_taxa(ps3, Family == "Enterobacteriaceae")
taxa_names(ps3.carson) <- paste0("SV", seq(ntaxa(ps3.carson)),"-",tax_table(ps3.carson)[,7])

gg.entero <- plot_tree(ps3.carson, ladderize="left",  color="psyllid_spp", label.tips="taxa_names") +   theme(legend.position = "none") +     #scale_colour_manual(values=col) +
  ggtitle("Enterobacidae 16S gene tree")

#, label.tips="psyllid_spp

```

## Secondary symbionts

Subset to clades within the 16S gene tree, do a paco analysis with that clade. Return test statistics.



### Fit neutral model

Cluster at a rate that approximates genus (95%?) Before fititng neutral model
Loop through those that are above the predicitons, doing a mantel test for each

Specifically, points above the prediction represent taxa that are found more frequently than expected, suggesting that they are actively being maintained and selected for by the host, while points found below the prediction represent taxa found less frequently than expected, suggesting that they are either selected against by the host or are especially dispersal limited. 


```{r}
library(reltools)
library(minpack.lm)
library(Hmisc)

ps.neutral <- speedyseq::tax_glom(ps2, taxrank="Genus")

ps.neutral <- tip_glom(ps2, h=0.3)

spp <- otu_table(ps.neutral)@.Data
spp.out <- fit_sncm(spp, pool=spp, taxon=data.frame(tax_table(ps.neutral)))

df <- data.frame(t(table(spp.out$predictions$fit_class)))
df <- df[,c(2,3)]
colnames(df) <- c("Prediction", "AVS Abundance")
  
library(ggrepel)
gg.neutral <- ggplot(data = spp.out$predictions) + 
  geom_point(aes(x = log(p), y = freq, fill = fit_class),
        shape = 21, color = "black", size = 2, alpha = 0.75) +
  scale_fill_manual(name = "Prediction", 
        values = c(`Above prediction` = "cyan4", 
            `As predicted` = "blue", `Below prediction` = "goldenrod", 
            `NA` = "white"), breaks = c("Above prediction", 
            "As predicted", "Below prediction", 
            "NA"), labels = c(paste0("Above prediction (", 
            round((df[1, 2]/spp.out$fitstats$Richness) * 
              100, 1), "%)"), paste0("As predicted (", 
            round((df[2, 2]/spp.out$fitstats$Richness) * 
              100, 1), "%)"), paste0("Below Prediction (", 
            round((df[3, 2]/spp.out$fitstats$Richness) * 
              100, 1), "%)"), paste0("NA (", df[4, 2], ")"))) +
  geom_line(aes(x = log(p), y = freq.pred), color = "blue") + 
  geom_line(aes(x = log(p), y = pred.lwr), color = "blue", linetype = "dashed") + 
  geom_line(aes(x = log(p), y = pred.upr), color = "blue", linetype = "dashed") +
  theme_bw() +
  #geom_label_repel(data = (spp.out$predictions %>% filter(fit_class == "Above prediction")), aes(x = log(p), y = freq,  label=Genus))+
  xlab("log(Mean Relative Abundance)") +
  ylab("Frequency") +
  annotate("text", x = mean(log(spp.out$predictions$p), na.rm = TRUE),
           y = 0.95, size = 5, label =  paste("r^2 ==", round(spp.out$fitstats$Rsqr, 4)), parse = TRUE) +  
  annotate("text", x = mean(log(spp.out$predictions$p), na.rm = TRUE),
           y = 0.9, size = 5, label =  paste("m ==", round(spp.out$fitstats$m,  4)), parse = TRUE)


```


## Community assembly

SEE: https://www.nature.com/articles/s41467-019-10191-3#Sec10

Our findings that host evolutionary history significantly explain microbiome diversity indicate that environmental filtering plays a substantial role in microbial community assembly. 

n order to further test this notion and to assess how environmental filtering may differ among host clades, we utilized two ecophylogenetics analyses: mean phylogenetic distance (MPD) and mean nearest taxon distance (MNTD). These tests assess the degree of phylogenetic clustering within each sample (host) relative to a permuted null model. Assuming phylogenetic niche conservatism (i.e., closely related taxa overlap along niche axes), then host diet or gut physiology may select for phylogenetically clustered taxa with overlapping niches, while in the absence of such strong selection, competition via niche conservatism would lead to phylogenetic overdispersion40. Phylogenetic overdispersion may also result from facilitation (i.e., beneficial microbe–microbe interactions), such as when distantly related taxa form consortia to break down complex plant polymers40. MPD is more sensitive to overall patterns of phylogenetic clustering and evenness, while MNTD is more sensitive to patterns at the tree tips41.




```{r session-info}
# Display current R session information
sessionInfo()
```
