---
title: "Psyllid microbiome"
subtitle: "Statistical analysis"
author: "Alexander Piper"
date: "`r Sys.Date()`"
output:
  html_document:
    includes:
      after_body: footer.html
    highlighter: null
    theme: "flatly"
    code_download: TRUE
    toc: TRUE
    toc_float:
      collapsed: FALSE
      smooth_scroll: TRUE
    df_print: paged
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Knitr global setup - change eval to true to run code
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, message=FALSE, error=FALSE,fig.show = "hold", fig.keep = "all")
opts_chunk$set(dev = 'png')
```

# Load packages 

```{r Load packages, eval=TRUE, warning=FALSE, message=FALSE, error=FALSE}
#Set required packages
.cran_packages <- c("tidyverse", 
                    "patchwork", 
                    "vegan", 
                    "seqinr",
                    "ape", 
                    "sp",
                    "data.table", 
                    "RColorBrewer",
                    "ggtree", 
                    "castor", 
                    "picante",
                    "phylosignal", 
                    "adephylo",
                    "paco",
                    "phytools",
                    "ecodist")
.bioc_packages <- c("dada2",
                    "phyloseq", 
                    "DECIPHER",
                    "Biostrings",
                    "ShortRead", 
                    "philr",
                    "ALDEx2")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}

sapply(c(.cran_packages,.bioc_packages), require, character.only = TRUE)

# Github packages
devtools::install_github("alexpiper/taxreturn")
library(taxreturn)
devtools::install_github("alexpiper/seqateurs")
library(seqateurs)
devtools::install_github("mikemc/speedyseq")
library(speedyseq)
#devtools::install_github('ggloor/CoDaSeq/CoDaSeq')
library(CoDaSeq)
devtools::install_github("easystats/report")
library(report)

#Source internal functions
source('R/BDTT.R')
source('R/phylosymbiosis.R')
source('R/helper_functions.R')
options(stringsAsFactors = FALSE)
```

# Read in phyloseq object

```{r read in phyloseq}
ps2 <- readRDS("output/rds/ps2.rds")
tree <- read.tree("output/phytree.nwk")

phy_tree(ps2) <- tree

# export filtered
dir.create("output/otu_tables/filtered")
seqateurs::summarise_taxa(ps2, "species", "SampleID") %>%
  spread(key="SampleID", value="totalRA") %>%
  write.csv(file = "output/otu_tables/filtered/filtered_spp_sum.csv")

seqateurs::summarise_taxa(ps2, "genus", "SampleID") %>%
  spread(key="SampleID", value="totalRA") %>%
  write.csv(file = "output/otu_tables/filtered/filtered_gen_sum.csv")

#Rename taxa - only keep first 30 characters
taxa_names(ps2) <- substr(paste0("SV", seq(ntaxa(ps2)),"-",tax_table(ps2)[,7]), 1,30)
```

## Create species merged table

```{r mergedspp}
# Merge species for beta diversity
ps.sppmerged <- ps2 %>%
    merge_samples(group = "psyllid_spp", fun=mean)

#This loses the sample metadata - Need to add it agian
sample_data(ps.sppmerged) <- sample_data(ps2) %>%
  filter(!duplicated(psyllid_spp)) %>%
  magrittr::set_rownames(.$psyllid_spp)

ps3 <- ps.sppmerged
```

## Read in psyllid phylogeny
```{R psyllid phylogeny}
psyllid_tree <- read.tree(text=readLines("sample_data/psyllid_beast_tree.nwk"))

# Match names with alpha diversity
psyllid_tree$tip.label <- psyllid_tree$tip.label %>%
  str_squish() %>%
  str_replace_all(pattern="\\.", replacement=" ") %>%
  str_replace_all(pattern="Acizzia hakae", replacement="Acizzia hakeae") %>%
  str_replace_all(pattern="POLLENISLAND", replacement="POLLEN ISLAND") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiae$", replacement="Ctenarytaina fuchsia A") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeB", replacement="Ctenarytaina fuchsia B") %>%
  str_replace_all(pattern="Ctenarytaina fuchsiaeC", replacement="Ctenarytaina fuchsia C") %>%
  str_replace_all(pattern="Ctenarytaina clavata", replacement="Ctenarytaina clavata sp ") %>%
  str_replace_all(pattern="Ctenarytaina clavata sp $", replacement="Ctenarytaina clavata sp A") %>%
  str_replace_all(pattern="Ctenarytaina sp$", replacement="Ctenarytaina sp ") %>%
  str_replace_all(pattern="Ctenarytaina spA", replacement="Ctenarytaina sp A") %>%
    str_replace_all(pattern="Ctenarytaina spB", replacement="Ctenarytaina sp B") %>%
  str_replace_all(pattern="Ctenarytaina unknown", replacement="Ctenarytaina insularis") %>%  
  str_replace_all(pattern="Psylla apicalisA", replacement="Psylla frodobagginsi") %>%
  str_replace_all(pattern="Psylla apicalisB", replacement="Psylla apicalis") %>%
  str_replace_all(pattern="carmichaeliae", replacement="carmichaeliae ") %>%
  str_replace_all(pattern="Trioza sp", replacement="Trioza sp ") %>%
  str_replace_all(pattern="Trioza acutaB", replacement="Trioza acuta B") %>%
  str_replace_all(pattern="Trioza gourlay", replacement="Trioza gourlayi") %>%
  str_replace_all(pattern="BRENDAMAY", replacement="BRENDA MAY") %>%
  str_replace_all(pattern="PRICES", replacement="PRICES VALLEY") %>%  
  str_replace_all(pattern="Acizzia sp", replacement="Acizzia errabunda") %>% 
  str_replace_all(pattern=" ", replacement="_") %>%
  trimws(which="right")

# Subset to only those in sample data
psyllid_tree$tip.label[!psyllid_tree$tip.label %in% sample_data(ps2)$psyllid_spp]
pruned.tree <- drop.tip(psyllid_tree, psyllid_tree$tip.label[!psyllid_tree$tip.label %in% sample_data(ps2)$psyllid_spp] )

```


## Summary statistics

```{r sum taxa}
# N unique species and samples
speedyseq::psmelt(ps2) %>%
  summarise(ntaxa= n_distinct(psyllid_spp), n_samples = n_distinct(Sample_Name), n_hostplants = n_distinct(hostplant_spp))

# Spread of reads
speedyseq::psmelt(ps2) %>%
  group_by(Sample_Name) %>%
  summarise(Abundance = sum(Abundance)) %>%
  ungroup() %>%
  summarise(mean = mean(Abundance), 
            se = sd(Abundance)/sqrt(length(Abundance)),
            max = max(Abundance),
            min = min(Abundance))

# Spread of ASVs
speedyseq::psmelt(ps2) %>%
  group_by(Sample_Name) %>%
  dplyr::filter(Abundance > 0) %>%
  summarise(counts = n_distinct(OTU)) %>%
  ungroup() %>%
  summarise(mean = mean(counts), 
            se = sd(counts)/sqrt(length(counts)),
            max = max(counts),
            min = min(counts))

#Fraction of reads assigned to each taxonomic rank
speedyseq::psmelt(ps2) %>%
  gather("Rank","Name",rank_names(ps2)) %>%
  group_by(Rank) %>% 
  mutate(Name = replace(Name, str_detect(Name, "__"),NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  dplyr::summarise(Reads_classified = sum(Abundance * !is.na(Name))) %>%
  mutate(Frac_reads = Reads_classified / sum(sample_sums(ps2))) %>%
  mutate(Rank = factor(Rank, rank_names(ps2))) %>%
  arrange(Rank)

#Fraction of ASV's assigned to each taxonomic rank
tax_table(ps2) %>%
  as("matrix") %>%
  as_tibble(rownames="OTU") %>%
  gather("Rank","Name",rank_names(ps2)) %>%
  group_by(Rank) %>%
  mutate(Name = replace(Name, str_detect(Name, "__"), NA)) %>% # This line turns the "__" we added to lower ranks back to NA's
  dplyr::summarise(OTUs_classified = sum(!is.na(Name))) %>%
  mutate(Frac_OTUs = OTUs_classified / ntaxa(ps2)) %>%
  mutate(Rank = factor(Rank, rank_names(ps2))) %>%
  arrange(Rank)

# Unique taxa at each rank
speedyseq::psmelt(ps2) %>%
  dplyr::select(rank_names(ps2)) %>%
  pivot_longer(everything(),
               names_to = "Rank",
               values_to = "value") %>%
  mutate(value = case_when(
    str_detect(value, "__") ~ as.character(NA),
    !str_detect(value, "__") ~ value
  )) %>%
  drop_na() %>%
  group_by(Rank) %>%
  summarise_all(funs(n_distinct)) %>%
  mutate(Rank = factor(Rank, rank_names(ps2))) %>%
  arrange(Rank)

# Each different phylum ranked by its overall relative abundance

# Transform to per sample relative abundance, then transform to whole dataset relative abundance

```
## Prevalence assesment

View prevalence of different phyla
```{r prevalence-assessment}
# Calculate taxon prevalence across the data set at OTU level
prevdf <- apply(X = otu_table(ps2), MARGIN = ifelse(taxa_are_rows(ps2), yes = 1, no = 2), FUN = function(x){sum(x > 0)})
prevdf <- data.frame(Prevalence = prevdf, 
                     TotalAbundance = taxa_sums(ps2),
                     tax_table(ps2))
#Prevalence plot
gg.prev <- subset(prevdf, phylum %in% get_taxa_unique(ps2, "phylum")) %>%
  ggplot(aes(TotalAbundance, Prevalence / nsamples(ps.merged),color=order)) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~phylum) +
  theme(legend.position="none") +
  ggtitle("Phylum Prevalence in All Samples\nColored by Order")

pdf(file="figs/prevalence.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.prev)
try(dev.off(), silent=TRUE)
  
  
# Prevalecne at phylum
ps.phylum <- speedyseq::tax_glom(ps2, taxrank="phylum")
prevdf_phylum <- apply(X = otu_table(ps.phylum ), MARGIN = ifelse(taxa_are_rows(ps.phylum ), yes = 1, no = 2), FUN = function(x){sum(x > 0)})
prevdf_phylum <- data.frame(Prevalence = prevdf_phylum, 
                     TotalAbundance = taxa_sums(ps.phylum),
                     tax_table(ps.phylum)) %>%
  dplyr::mutate(RA = TotalAbundance / sum(TotalAbundance)) %>%
  remove_rownames() %>%
  magrittr::set_rownames(.$phylum) %>%
  select(-rank_names(ps.phylum))

# Prevalence within Proteobacteria
ps.prot <- subset_taxa(ps2, phylum=="Proteobacteria") %>%
          speedyseq::tax_glom(taxrank="order")
prevdf_prot <- apply(X = otu_table(ps.prot ), MARGIN = ifelse(taxa_are_rows(ps.prot ), yes = 1, no = 2), FUN = function(x){sum(x > 0)})
prevdf_prot <- data.frame(Prevalence = prevdf_prot, 
                     TotalAbundance = taxa_sums(ps.prot),
                     tax_table(ps.prot)) %>%
  dplyr::mutate(RA = TotalAbundance / sum(TotalAbundance)) %>%
  remove_rownames() %>%
  magrittr::set_rownames(.$order) %>%
  select(-rank_names(ps.prot))
  
# Genus Prevalence
ps.gen <- speedyseq::tax_glom(ps2, taxrank="genus") 
prevdf_gen <- apply(X = otu_table(ps.gen ), MARGIN = ifelse(taxa_are_rows(ps.gen ), yes = 1, no = 2), FUN = function(x){sum(x > 0)})
prevdf_gen <- data.frame(Prevalence = prevdf_gen, 
                     TotalAbundance = taxa_sums(ps.gen),
                     tax_table(ps.gen)) %>%
  dplyr::mutate(RA = TotalAbundance / sum(TotalAbundance)) %>%
  remove_rownames() %>%
  magrittr::set_rownames(.$genus) %>%
  select(-rank_names(ps.gen))

# Genus Prevalence across species rather than specimens
ps.gen <- speedyseq::tax_glom(ps3, taxrank="genus") 
prevdf_gen <- apply(X = otu_table(ps.gen ), MARGIN = ifelse(taxa_are_rows(ps.gen ), yes = 1, no = 2), FUN = function(x){sum(x > 0)})
prevdf_gen <- data.frame(Prevalence = prevdf_gen, 
                     TotalAbundance = taxa_sums(ps.gen),
                     tax_table(ps.gen)) %>%
  dplyr::mutate(RA = TotalAbundance / sum(TotalAbundance)) %>%
  remove_rownames() %>%
  magrittr::set_rownames(.$genus) %>%
  select(-rank_names(ps.gen))

# CHECK WHICH TAXA CARSONELLA IS ABSENT IN
speedyseq::tax_glom(ps2, taxrank="genus")  %>%
  microbiome::transform(transform="compositional") %>%
  speedyseq::psmelt() %>%
  filter(genus =="Candidatus Carsonella") %>% 
  summarise(mean = mean(Abundance), 
            se = sd(Abundance)/sqrt(length(Abundance)),
            sd = sd(Abundance),
            max = max(Abundance),
            min = min(Abundance))
```

## Alpha diversity metrics

```{r alpha diversity}
dir.create("output/alpha")
# Get richness measures
richness <- phyloseq::estimate_richness(ps2, measures=c("Shannon")) %>%
  rownames_to_column("Sample_Name") %>%
  mutate(Sample_Name = Sample_Name %>% 
           str_remove("^X") %>%
           str_replace_all("\\.", " "))

#Set number of randomisations for calculating significance
# Calculate Faith's PD-index & Species richness - with Standard errors
sespd <- picante::ses.pd(as(phyloseq::otu_table(ps2), "matrix"),  phyloseq::phy_tree(ps2), null.model = "taxa.labels", include.root = F, runs = 99)

# Join together
div_table <- sespd %>%
  rownames_to_column("Sample_Name") %>%
  dplyr::select(Sample_Name, ntaxa, pd.obs) %>%
  dplyr::rename(pd = pd.obs, alpha = ntaxa) %>%
  left_join(richness, by="Sample_Name") %>%
  left_join(sample_data(ps2) %>% 
              as.data.frame() %>%
              filter(!duplicated(Sample_Name)) %>%
              dplyr::select(Sample_Name, psyllid_spp),
            by = "Sample_Name") 
# Summarise means
div_table %>%
  summarise_if(is.numeric, mean)

# Difference between factors ANOVA
report::report(aov(alpha ~psyllid_spp, data=div_table))
report::report(aov(Shannon ~psyllid_spp, data=div_table))
report::report(aov(pd ~psyllid_spp, data=div_table))
# Association with phylogeny

dat <- div_table  %>%
  filter(psyllid_spp %in% pruned.tree$tip.label) %>% #Subset to common 
  group_by(psyllid_spp) %>%
  dplyr::select(-Sample_Name) %>%
  summarise_all(mean) %>%
  arrange(match(psyllid_spp, pruned.tree$tip.label)) %>%
  as.data.frame() %>%
  magrittr::set_rownames(.$psyllid_spp) %>%
  dplyr::select(-psyllid_spp)

# Add positive and negative controls
dat$random <- rnorm(length(dat$alpha), sd = 10) #Random association
dat$bm <- rTraitCont(pruned.tree) #Brownian motion

# Make phylosignal object and measure signal between univariate traits.
p4d <- phylobase::phylo4d(pruned.tree, dat)	
signal <- phylosignal::phyloSignal(p4d = p4d, methods = c("I", "Lambda", "K"), reps = 999)%>%
  as.data.frame() %>%
  rownames_to_column("measure")

# print phylogenetic signal
signal
write_csv(signal, "output/alpha/phylosignal.csv")

# Locate signal
lipa <- lipaMoran(p4d, reps=999)
lipa.p4d <- lipaMoran(p4d, as.p4d = TRUE, reps=999)
barplot.phylo4d(lipa.p4d, bar.col = (lipa$p.value < 0.05) + 1, center = FALSE, scale = FALSE) + title("Non-rarefied")

#write out lipa
lipa_out <- cbind(lipa$lipa %>%
                    as.data.frame %>%
                    rename_all(funs(paste0(., "_stat"))),
                  lipa$p.value %>%
                    as.data.frame %>%
                    rename_all(funs(paste0(., "_pval")))
                               )%>%
  rownames_to_column("psyllid_spp")
write_csv(lipa_out, "output/alpha/lipa.csv")                  
                  
```

## Rarefied

See if the pattern holds even with rarefaction to lowest sample

```{r Phylosignal}
# Rarefied richness
ps2_rare <- rarefy_even_depth(ps2, sample.size = min(sample_sums(ps2)),
  rngseed = 666, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)

# Get richness measures
richness_rare <- phyloseq::estimate_richness(ps2_rare, measures=c("Shannon")) %>%
  rownames_to_column("Sample_Name") %>%
  mutate(Sample_Name = Sample_Name %>% 
           str_remove("^X") %>%
           str_replace_all("\\.", " "))

#Set number of randomisations for calculating significance
# Calculate Faith's PD-index & Species richness - with Standard errors
sespd_rare <- picante::ses.pd(as(phyloseq::otu_table(ps2_rare), "matrix"),  phyloseq::phy_tree(ps2_rare), null.model = "taxa.labels", include.root = F, runs = 99)

# Join together
div_table_rare <- sespd_rare %>%
  rownames_to_column("Sample_Name") %>%
  dplyr::select(Sample_Name, ntaxa, pd.obs) %>%
  dplyr::rename(pd = pd.obs, alpha = ntaxa) %>%
  left_join(richness_rare, by="Sample_Name") %>%
  left_join(sample_data(ps2) %>% 
              as.data.frame() %>%
              filter(!duplicated(Sample_Name)) %>%
              dplyr::select(Sample_Name, psyllid_spp),
            by = "Sample_Name") 
# Summarise means
div_table_rare %>%
  summarise_if(is.numeric, mean)


dat <- div_table_rare  %>%
  filter(psyllid_spp %in% pruned.tree$tip.label) %>% #Subset to common 
  group_by(psyllid_spp) %>%
  dplyr::select(-Sample_Name) %>%
  summarise_all(mean) %>%
  arrange(match(psyllid_spp, pruned.tree$tip.label)) %>%
  as.data.frame() %>%
  magrittr::set_rownames(.$psyllid_spp) %>%
  dplyr::select(-psyllid_spp)

# Add positive and negative controls
dat$random <- rnorm(length(dat$alpha), sd = 10) #Random association
dat$bm <- rTraitCont(pruned.tree) #Brownian motion

# Make phylosignal object and measure signal between univariate traits.
p4d <- phylobase::phylo4d(pruned.tree, dat)	
signal_rare <- phylosignal::phyloSignal(p4d = p4d, methods = c("I", "Lambda", "K"), reps = 999) %>%
  as.data.frame() %>%
  rownames_to_column("measure")

# print phylogenetic signal
signal_rare
write_csv(signal_rare, "output/alpha/phylosignal_rarefied.csv")

# Locate signal
lipa <- lipaMoran(p4d, reps=999)
lipa.p4d <- lipaMoran(p4d, as.p4d = TRUE, reps=999)
barplot.phylo4d(lipa.p4d, bar.col = (lipa$p.value < 0.05) + 1, center = FALSE, scale = FALSE) + title("Rarefied")

#write out lipa
lipa_out_rare <- cbind(lipa$lipa %>%
                    as.data.frame %>%
                    rename_all(funs(paste0(., "_stat"))),
                  lipa$p.value %>%
                    as.data.frame %>%
                    rename_all(funs(paste0(., "_pval")))
                               )%>%
  rownames_to_column("psyllid_spp")
write_csv(lipa_out_rare, "output/alpha/lipa_rarefied.csv")    
```	

# Beta diversity

## Microbe distances
```{r distlist}
# Test filtered
filterfun1 <- function(x){
  x[(x / sum(x)) < (1e-4)] <- 0
  return(x)
}
ps2_filt  <- transform_sample_counts(ps2, fun = filterfun1) %>%
  filter_taxa(function(x) mean(x) > 0, TRUE) #Drop missing taxa from table

print(paste0((ntaxa(ps2)-ntaxa(ps2_filt)), " taxa under threshold removed"))

#ps2_dist <- ps2
ps2_dist <- ps2_filt

# Get OTU tables
otutab <- otu_table(ps2_dist)
#Impute zeroes for compositional distances
otutab_n0 <- as.matrix(zCompositions::cmultRepl(otutab, method="BL", output="p-counts"))
#Root phylogenetic tree
phy_tree(ps2_dist) <- multi2di(phy_tree(ps2_dist))

#Calculate different distance metrics
metrics <- c("Bray", "Jaccard", "Aitchison", "Unifrac", "WUnifrac")  #"Philr",
distlist <- vector("list", length=length(metrics))
names(distlist) <- metrics

distlist$Jaccard <- as.matrix(vegdist(otutab, method="jac",binary = T))
distlist$Bray <- as.matrix(vegdist(otutab, method="bray"))
distlist$Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_n0), method="euclidean"))
#distlist$Philr <- as.matrix(vegdist(philr::philr(otutab_n0, phy_tree(ps2),
#                                                part.weights='enorm.x.gm.counts',
#                                                ilr.weights='blw.sqrt'), method="euclidean"))
distlist$Unifrac <- as.matrix(phyloseq::UniFrac(ps2_dist, weighted=FALSE, parallel = TRUE))
distlist$WUnifrac <- as.matrix(phyloseq::UniFrac(ps2_dist, weighted=TRUE, parallel = TRUE))

```


## Adonis & Betadisper

```{r Adonis}
# Adonis test
metadata <- sample_data(ps2) %>%
  as_data_frame()
adonis_results <- distlist %>%
  purrr::map(function(x) {
    bind_rows(
    broom::tidy(adonis(x~psyllid_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1),
    broom::tidy(adonis(x~hostplant_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1)
    )
})  %>%
  bind_rows(.id="dist")

# Check homogeneity
betadisper_results <- distlist %>%
  purrr::map(function(x) {
    y <- as.dist(x[metadata$Sample_Name, metadata$Sample_Name])
  bind_rows(
    as.data.frame(permutest(vegan::betadisper(y, metadata$psyllid_spp))$tab) %>%
      dplyr::slice(1) %>% 
      mutate(term="psyllid_spp"),
    as.data.frame(permutest(vegan::betadisper(y, metadata$hostplant_spp))$tab) %>%
      dplyr::slice(1) %>% 
      mutate(term="hostplant_spp"),
  )
})  %>%
  bind_rows(.id="dist")

  y <- as.dist(distlist[["Aitchison"]][metadata$Sample_Name, metadata$Sample_Name])

test <- permutest(vegan::betadisper(y, metadata$psyllid_spp))

dir.create("output/beta")
write_csv(adonis_results, "output/beta/adonis_fulldata.csv")
write_csv(betadisper_results, "output/beta/adonis_fulldata.csv")
```

## Same tree dissimilarities

```{r Avg Dissimiarity}
hostplant_metadata <- metadata %>% mutate(ingroup = case_when(
  Sample_Name %in% c("94","107", "113","93","106", "112") ~ "fraxini-fraxinicola",
  Sample_Name %in% c("200big", "201big", "200small", "201small") ~ "apicalis-frodobagginsi",
  TRUE ~ "other"
  ))

broom::tidy(adonis(distlist$Aitchison~ingroup + psyllid_spp, method="euclidean",
                   data=hostplant_metadata)$aov.tab)

pairwise.adonis2(distlist$Aitchison~ingroup + psyllid_spp, method="euclidean",
                   data=hostplant_metadata)
```

## Barplot

```{r barplot}
# Plot tree
p <- ggtree(pruned.tree) + geom_tiplab(align=TRUE) + geom_nodelab(geom='label') +
    scale_x_continuous(expand=c(0, 0.1)) + theme_tree2()

# Plot bar
ps3_bar <- ps3 %>%
  speedyseq::tax_glom(taxrank = "order") %>%           # agglomerate at Order level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  speedyseq::psmelt() %>%
  mutate(plotlabel = phylum) %>%
  mutate(plotlabel = case_when(
    Abundance >= 0.01 & phylum=="Proteobacteria" ~ paste0("P - ", order), # Change this to whatever taxrank we want
    Abundance >= 0.01 & !phylum=="Proteobacteria"~ phylum ,
    Abundance < 0.01 ~ "NA"
    )) %>%
  dplyr::na_if("NA") %>%
  dplyr::select(psyllid_spp, plotlabel, phylum, order, Abundance) %>%
  left_join(p$data %>%
              as_data_frame %>%
              dplyr::filter(isTip) %>%
              dplyr::select(y, label) %>%
              dplyr::rename(psyllid_spp=label)) 

gg.bar <- ggplot(ps3_bar, aes(x=y, y=Abundance, fill=plotlabel)) +
  geom_col()  + 
  coord_flip()+
  scale_fill_manual(values=colorRampPalette(brewer.pal(9, "Set1"))(length(unique(ps3_bar$plotlabel))-1), na.value="grey") + 
    theme_classic()   +
    theme(legend.position = "bottom",
      panel.grid.major.x = element_line(colour="grey92", size=0.5, linetype="dashed"),
      strip.background = element_rect(fill = "grey92", 
                    colour = "black", size = 1),
      axis.text.y = element_blank(),
      axis.ticks.y=element_blank(),
      axis.title.y=element_blank()) +
  scale_y_continuous(expand=c(0,0))

# Make richness plots
gg.rich <-  div_table %>%
  group_by(psyllid_spp) %>%
  summarise(alpha=mean(alpha), pd=mean(pd)/10e+7, Shannon=mean(Shannon)) %>%
  ungroup() %>%
  pivot_longer(-psyllid_spp, names_to="measure",
               values_to = "value")  %>%
  left_join(lipa_out %>% 
              dplyr::select(psyllid_spp, alpha_pval, pd_pval, Shannon_pval) %>%
              pivot_longer(-psyllid_spp,
                           names_to="measure",
               values_to = "pval") %>%
              mutate(measure = str_remove(measure, "_pval"))) %>%
  left_join(p$data %>%
              as_data_frame %>%
              dplyr::filter(isTip) %>%
              dplyr::select(y, label) %>%
              dplyr::rename(psyllid_spp=label)) %>%
  mutate_at(vars(measure), funs(factor(., levels=c("alpha","Shannon","pd")))) %>%
  ggplot(aes(x=y, y=value, fill=pval<0.05)) + 
    geom_col() +
    facet_grid(~measure, scales="free") + 
    coord_flip()+
    theme_classic()   +
    theme(legend.position = "bottom",
    panel.grid.major.x = element_line(colour="grey92", size=0.5, linetype="dashed"),
    strip.background = element_rect(fill = "grey92", 
                  colour = "black", size = 1),
    axis.text.y = element_blank(),
    axis.ticks.y=element_blank(),
    axis.title.y=element_blank(),
    axis.text.x = element_text(angle=45, hjust=1)) +
    scale_fill_manual(values=c("darkgray", "darkred")) +
    scale_x_continuous(breaks=scales::pretty_breaks(n=1))

## Collection_hist
gg.spp <- as_tibble(sample_data(ps2)) %>%
  dplyr::rename(label = psyllid_spp) %>%
  group_by(label) %>%
  summarise(n_species = n()) %>%
  left_join(p$data %>%
  filter(isTip) %>% select(c(label, y))) %>%
  filter(!is.na(y)) %>%
  ggplot(aes(x=y, y=1, fill=n_species)) +
  geom_tile() +
  coord_flip() +
  theme_void() +
  theme(legend.position = "bottom") +
  scale_fill_distiller(palette = "Reds", direction=1)

#Arrange
Fig1 <- p + gg.spp + gg.bar + gg.rich + plot_layout(nrow=1, widths=c(1,0.08,2,0.6))  

pdf(file="figs/Beta.pdf", width = 8, height = 11 , paper="a4")
  plot(Fig1)
try(dev.off(), silent=TRUE)
```

## Phylosymbiosis

### Prepare distance matrices
```{r prepare distances}
## Psyllid phylogeny cophenetic distance matrix
phylo.dist <-cophenetic(pruned.tree) %>%
   sqrt() %>%
  as.data.frame() %>%
  rownames_to_column("psyllid_spp.x") %>%
  pivot_longer(cols=-psyllid_spp.x,
               names_to="psyllid_spp.y",
               values_to = "dist") %>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, psyllid_spp) %>%
              dplyr::rename(psyllid_spp.x = psyllid_spp, Sample_Name.x = Sample_Name),
            by="psyllid_spp.x")%>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, psyllid_spp) %>%
              dplyr::rename(psyllid_spp.y = psyllid_spp, Sample_Name.y = Sample_Name),
            by="psyllid_spp.y") %>%
  filter(!is.na(dist))%>%
  dplyr::select(-psyllid_spp.y, -psyllid_spp.x) %>%
  pivot_wider(names_from = Sample_Name.y, values_from = dist) %>%
  magrittr::set_rownames(.$Sample_Name.x) %>%
  dplyr::select(-Sample_Name.x)%>%
  as.matrix()

plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.dist <- cophenetic(plant.tree) %>%
  sqrt() %>%
  as.data.frame() %>%
  rownames_to_column("hostplant_spp.x") %>%
  pivot_longer(cols=-hostplant_spp.x,
               names_to="hostplant_spp.y",
               values_to = "dist") %>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, hostplant_spp) %>%
              dplyr::rename(hostplant_spp.x = hostplant_spp, Sample_Name.x = Sample_Name),
            by="hostplant_spp.x")%>%
  right_join(sample_data(ps2) %>%
              as_tibble() %>%
              dplyr::select(Sample_Name, hostplant_spp) %>%
              dplyr::rename(hostplant_spp.y = hostplant_spp, Sample_Name.y = Sample_Name),
            by="hostplant_spp.y") %>%
  filter(!is.na(dist))%>%
  dplyr::select(-hostplant_spp.y, -hostplant_spp.x) %>%
  pivot_wider(names_from = Sample_Name.y, values_from = dist) %>%
  magrittr::set_rownames(.$Sample_Name.x) %>%
  dplyr::select(-Sample_Name.x)%>%
  as.matrix()

# Spatial distance matrix
envData <- sample_data(ps2) %>%
  as_data_frame() %>%
  dplyr::select(Sample_Name, lat, long) %>%
  tibble::column_to_rownames("Sample_Name") %>%
  drop_na()

spat.dist <- spDists(as.matrix(envData), longlat=TRUE) %>%
  as.data.frame() %>%
  magrittr::set_rownames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  magrittr::set_colnames(rownames(envData) %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  rownames_to_column("SampleID") %>%
  mutate(SampleID = SampleID %>% str_replace(pattern="\\_S(.*)$",replacement="") %>% make.unique()) %>%
  column_to_rownames("SampleID") %>%
  as.matrix()
```

### Whole dataset

But also see how this paper has done it with PACO https://www.nature.com/articles/s41467-019-10191-3.pdf Using random subsampling? Then pull out the most associated also file:///C:/Users/ap0y/Dropbox/workprojects/Papers/10.1002_ecy.1955.pdf

```{r phylosymbiosis}
set.seed(909)
dir.create("output/phylosymbiosis")

# Matrix correlations
#only use samples present in all
subsample <- Reduce(intersect, list(rownames(otu_table(ps2)), colnames(phylo.dist), colnames(plant.dist), colnames(spat.dist)))

# Mantel test
mantel_results <- distlist %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               subsample=subsample, type="mantel")
  }) %>%
  bind_rows(.id="dist") 

write_csv(mantel_results %>% dplyr::select(-pval1,-pval2),  "output/phylosymbiosis/mantel_fulldata.csv")#only keep two sided P values

# Partial Mantel Test
pmantel_results <- distlist %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               subsample=subsample, type="partial")
  }) %>%
  bind_rows(.id="dist") 

write_csv(pmantel_results %>% dplyr::select(-pval1,-pval2), "output/phylosymbiosis/pmantel_fulldata.csv")

## Plot circular barplot
gg.mantels <- bind_rows(mantel_results, pmantel_results) %>%
              dplyr::mutate(dist1 = case_when(
                str_detect(dist1, "plant.dist") ~ "Hostplant phylogeny",
                str_detect(dist1, "phylo.dist") ~ "Psyllid phylogeny",                
                str_detect(dist1, "spat.dist") ~ "Spatial distance"                   
              )) %>%
  filter(dist == "Aitchison") %>%
              ggplot(aes(x=dist1, y=mantelr, fill=dist1)) + 
  geom_bar(stat="identity", position="identity", alpha=1) +
  geom_errorbar(aes(ymin=`llim.2.5%`, ymax=`ulim.97.5%`), width=.1) +
  scale_fill_brewer(palette="Paired")  +
  facet_wrap(~type, ncol=1) +
  theme_classic() +
  theme(legend.position = "bottom",
        strip.background = element_rect(fill="grey85"))+
  guides(fill=guide_legend(ncol=1))

pdf(file="figs/bar_mantels.pdf", width = 5, height = 5 , paper="a4r")
  plot(gg.mantels)
try(dev.off(), silent=TRUE)
```

### Without Gammaproteobacteria

Probably dont need this bit if we are lookign at contribution to cophylogeny through paco
```{r phylosymbiosis no entero}
ps2_subset <- ps2 %>%
 subset_taxa(class != "Gammaproteobacteria") %>% #is this working?
 filter_taxa(function(x) mean(x) > 0, TRUE)#Drop missing taxa from table
ps2_subset <- prune_samples(sample_sums(ps2_subset) >0 , ps2_subset)
message(nsamples(ps2) - nsamples(ps2_subset), " Samples and ", ntaxa(ps2) - ntaxa(ps2_subset), " taxa Dropped")

# Get OTU tables
otutab_subset <- otu_table(ps2_subset)
#Impute zeroes for compositional distances
otutab_subset_n0 <- as.matrix(zCompositions::cmultRepl(otutab_subset, method="BL", output="p-counts"))
#Root phylogenetic tree
phy_tree(ps2_subset) <- multi2di(phy_tree(ps2_subset))

#Calculate different distance metrics
metrics <- c("Bray", "Jaccard", "Aitchison", "Unifrac", "WUnifrac") # "Philr"
distlist_subset <- vector("list", length=length(metrics))
names(distlist_subset) <- metrics

distlist_subset$Jaccard <- as.matrix(vegdist(otutab_subset, method="jac",binary = T))
distlist_subset$Bray <- as.matrix(vegdist(otutab_subset, method="bray"))
distlist_subset$Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_subset_n0, aitch=TRUE)
                                               , method="euclidean"))
#distlist_subset$Philr <- as.matrix(vegdist(philr::philr(otutab_n0, phy_tree(ps2),
#                                                part.weights='enorm.x.gm.counts',
#                                                ilr.weights='blw.sqrt'), method="euclidean"))
distlist_subset$Unifrac <- as.matrix(phyloseq::UniFrac(ps2_subset, weighted=FALSE, parallel = TRUE))
distlist_subset$WUnifrac <- as.matrix(phyloseq::UniFrac(ps2_subset, weighted=TRUE, parallel = TRUE))

# Adonis test
metadata <- sample_data(ps2_subset) %>%
  as_data_frame()
adonis_results <- distlist_subset %>%
  purrr::map(function(x) {
    bind_rows(
    broom::tidy(adonis(x~psyllid_spp, method="euclidean", data=metadata)$aov.tab) %>% dplyr::slice(1),
    broom::tidy(adonis(x~hostplant_spp, method="euclidean", data=metadata)$aov.tab)%>% dplyr::slice(1)
    )
})  %>%
  bind_rows(.id="dist")

write_csv(adonis_results, "output/beta/adonis_subset.csv")
# Matrix correlations
#only use samples present in all
subsample <- Reduce(intersect, list(rownames(otu_table(ps2_subset)), colnames(phylo.dist), colnames(plant.dist), colnames(spat.dist)))

# Mantel test
mantel_results <- distlist_subset %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               subsample=subsample, type="mantel")
  }) %>%
  bind_rows(.id="dist") 

write_csv(mantel_results %>% dplyr::select(-pval1,-pval2), "output/phylosymbiosis/mantel_subset.csv")

# Partial Mantel Test
pmantel_results <- distlist_subset %>%
  purrr::map(function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               subsample=subsample, type="partial")
  }) %>%
  bind_rows(.id="dist") 
write_csv(pmantel_results %>% dplyr::select(-pval1,-pval2), "output/phylosymbiosis/pmantel_subset.csv")
```

## BDTT

```{r timeslice}
#Look at depth categories in tree
hist(get_all_node_depths(phy_tree(ps2))) # PLOT THIS NEXT TO TREE

# Full slices - 2e+09 / 1e07
#Set slices every 100M years
slices=seq(0, 2e+09, 1e07)

treelist <- vector("list", length=length(slices))
for (i in 1:length(slices)){
  newps <- speedyseq::tree_glom(ps2, resolution = slices[i])
  treelist[[i]] <- phy_tree(newps)
}
class(treelist) <- "multiPhylo"
names(treelist) <- paste0("Timeslice = ",slices)
gg.BTDDtrees <- ggtree(treelist) + facet_wrap(~.id, scale="free") + theme_tree2()

Betas <- BDTT(ps2, slices, metrics=c("Bray", "Aitchison"), zeroes="Impute", cores=1, quiet=FALSE) #"Unifrac", "WUnifrac" , "Philr"

# Matrix correlations
#only use samples present in all
subsamples <- Reduce(intersect, list(rownames(otu_table(ps2)), colnames(phylo.dist), colnames(plant.dist), colnames(spat.dist)))

# Partial Mantel
BTDD_pm <- Betas %>%
  purrr::map_depth(.depth=2, function(x){
    run_mantel(x, dists=c("phylo.dist", "plant.dist", "spat.dist"),
               subsample=subsamples, type="partial")
  }) %>%
  purrr::map(bind_rows,.id="dist") %>%
  bind_rows(.id="slice") %>%
  mutate(Comparison = paste0("Microbiome ~ ",dist1," + ",dist2," + ",dist3),
         Mya = as.numeric(slice)/1e+6)

#Write out
write_csv(BTDD_pm %>% dplyr::select(-pval1,-pval2,-type),"output/phylosymbiosis/BTDD.csv")

gg.BTDD <- BTDD_pm %>%
  filter(dist=="Aitchison") %>%
  ggplot(aes(y=mantelr, x=Mya, colour=Comparison, group=factor(Comparison)))+ 
  geom_line(linetype="dashed") +
  # Add a ribbon with the confidence band
  geom_ribbon(aes(ymin = `llim.2.5%`, ymax = `ulim.97.5%`, fill=Comparison),colour=NA, alpha=0.3) +
  geom_line(data=BTDD_pm %>%
            filter(dist=="Aitchison") %>%
            mutate(mantelr = case_when (pval3 > 0.01 ~ as.double(NA), pval3 < 0.01 ~ mantelr)), linetype="solid") +
     theme_classic()   +
     theme(legend.position = "bottom",
           #panel.grid.major.y = element_line(colour="grey", size=0.5, linetype="dashed"),
           strip.background = element_rect(fill = "grey92", 
                                           colour = "black", size = 1)
     )  + 
  scale_color_brewer(palette="Paired")+
  scale_fill_brewer(palette="Paired")  +
  guides(colour=guide_legend(ncol=1))

# Would be nice to add the actual null model to this plot

pdf(file="figs/BTDD.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.BTDD)
try(dev.off(), silent=TRUE)
  
# output together
gg.allmantels <- gg.mantels + gg.BTDD  + 
  plot_layout(widths = c(1, 2))
  
pdf(file="figs/all_mantels.pdf",  width = 11, height = 8, paper="a4r")
  plot(gg.allmantels)
try(dev.off(), silent=TRUE)
```


## PCA plots
```{r PCA}
#Impute zeroes for compositional distances
otutab_n0 <- as.matrix(zCompositions::cmultRepl(otu_table(ps2), method="BL", output="p-counts"))

# Get aitchison distance
Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_n0), method="euclidean"))

#PCA 
r.pcx <- prcomp(Aitchison)
pc_samp <- data.frame(Sample_Name = rownames(r.pcx$x), r.pcx$x[, 1:2])%>%
  left_join(as_data_frame(sample_data(ps2)), by="Sample_Name")
# calculate percent variance explained for the axis labels
pc1 <- round(r.pcx$sdev[1]^2/sum(r.pcx$sdev^2),2)
pc2 <- round(r.pcx$sdev[2]^2/sum(r.pcx$sdev^2),2)
pc_ylab <- paste("PC1: ", pc1, sep="")
pc_xlab <- paste("PC2: ", pc2, sep="")
library(ggplot2)

## colour by psyllid phylogeny
dend <- as.dendrogram(force.ultrametric(pruned.tree))
membership <- as.data.frame(cutree(dend, k=11)) %>%
  rownames_to_column("psyllid_spp") %>%
  magrittr::set_colnames(c("psyllid_spp", "cluster"))

pca_phylo <- pc_samp %>%
  left_join(membership)
gg.pca <- ggplot(data=pca_phylo, aes(x=PC2, y=PC1, colour=as.factor(cluster))) + 
  geom_point(alpha=0.5, size=3) + #, shape=21, colour="black"
  #geom_point(data=pc_otu,aes(PC1, PC2)) +
  theme_classic() +
  scale_colour_brewer(palette="Paired") +
  geom_hline(yintercept = 0, linetype=2, alpha=0.5) +  
  geom_vline(xintercept = 0, linetype=2, alpha=0.5) +
  xlab(pc_xlab) + 
  ylab(pc_ylab) +
  theme(legend.position = "none") +
  scale_y_reverse(position = "right") 
  #coord_fixed(ratio=pc2/pc1) # Scale plot by variance explained
print(gg.pca )

p1 <- ggtree(pruned.tree) +
    scale_x_continuous(expand=c(0, 0.2)) + theme_tree2()

colours_p1 <- p1$data %>%
  left_join(membership %>%
  dplyr::rename(label = psyllid_spp)
    )
p1 <- p1 %<+% colours_p1  + 
  geom_tippoint(aes(colour=as.factor(cluster)))  + 
  geom_tiplab(aes(colour=as.factor(cluster)))+
  scale_colour_brewer(palette="Paired") 

gg.psyllid_pca <- p1 + gg.pca + plot_annotation(title="Psyllid phylogenetic distance")
gg.psyllid_pca

pdf(file="figs/psyllid_pca.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.psyllid_pca)
try(dev.off(), silent=TRUE)
  
# Colour by plant phylogeny
plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.tree  <- drop.tip(plant.tree, "Sophora_microphylla_-kowhai" )
plant.tree <- multi2di(plant.tree)

dend <- as.dendrogram(plant.tree )
#test <- color_branches(dend, k=12)
#plot(test)
membership <- as.data.frame(cutree(dend, k=12)) %>%
  rownames_to_column("hostplant_spp") %>%
  magrittr::set_colnames(c("hostplant_spp", "cluster"))

pca_phylo <- pc_samp %>%
  left_join(membership)
gg.pca <- ggplot(data=pca_phylo, aes(x=PC2, y=PC1, colour=as.factor(cluster))) + 
  geom_point(alpha=0.5, size=3) + #, shape=21, colour="black"
  #geom_point(data=pc_otu,aes(PC1, PC2)) +
  theme_classic() +
  scale_colour_brewer(palette="Paired") +
  geom_hline(yintercept = 0, linetype=2, alpha=0.5) +  
  geom_vline(xintercept = 0, linetype=2, alpha=0.5) +
  xlab(pc_xlab) + 
  ylab(pc_ylab) +
  theme(legend.position = "none") +
  scale_y_reverse(position = "right")
  #coord_fixed(ratio=pc2/pc1) # Scale plot by variance explained
print(gg.pca )

p2 <- ggtree(plant.tree) +
    scale_x_continuous(expand=c(0, 70)) + theme_tree2()

colours_p2 <- p2$data %>%
  left_join(membership %>%
  dplyr::rename(label = hostplant_spp)
    )
p2 <- p2 %<+% colours_p2  + 
  geom_tippoint(aes(colour=as.factor(cluster)))  + 
  geom_tiplab(aes(colour=as.factor(cluster)))+
  scale_colour_brewer(palette="Paired") 

gg.plant_pca <- p2 + gg.pca + plot_annotation(title="Plant phylogenetic distance")
gg.plant_pca

pdf(file="figs/plant_pca.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.plant_pca)
try(dev.off(), silent=TRUE)

# HC of spatial distance
  
dend <- hclust(as.dist(spat.dist), method="average")
#test <- color_branches(dend, k=12)
#plot(test)

membership <- as.data.frame(cutree(dend, k=12)) %>%
  rownames_to_column("Sample_Name") %>%
  magrittr::set_colnames(c("Sample_Name", "cluster"))

pca_phylo <- pc_samp %>%
  left_join(membership)
gg.pca <- ggplot(data=pca_phylo, aes(x=PC2, y=PC1, colour=as.factor(cluster))) + 
  geom_point(alpha=0.5, size=3) + #, shape=21, colour="black"
  #geom_point(data=pc_otu,aes(PC1, PC2)) +
  theme_classic() +
  scale_colour_brewer(palette="Paired") +
  geom_hline(yintercept = 0, linetype=2, alpha=0.5) +  
  geom_vline(xintercept = 0, linetype=2, alpha=0.5) +
  xlab(pc_xlab) + 
  ylab(pc_ylab) +
  theme(legend.position = "none") +
  scale_y_reverse(position = "right")
  #coord_fixed(ratio=pc2/pc1) # Scale plot by variance explained
print(gg.pca )

p3 <- ggtree(as.phylo(dend) )+
    scale_x_continuous(expand=c(0, 400)) + theme_tree2()

colours_p3 <- p3$data %>%
  left_join(membership %>%
  dplyr::rename(label = Sample_Name)
    )
p3 <- p3 %<+% colours_p3  + 
  geom_tippoint(aes(colour=as.factor(cluster)))  + 
  geom_tiplab(aes(colour=as.factor(cluster)))+
  scale_colour_brewer(palette="Paired") 

gg.spat_pca <- p3 + gg.pca + plot_annotation(title="Spatial Distance")
gg.spat_pca

pdf(file="figs/spatial_pca.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.spat_pca)
try(dev.off(), silent=TRUE)
```


# Cophylogeny

Due to the higher correlations with recent branches, we looked for patterns of cophylogeny

## Psyllid ~ Microbe

```{r PACO}
#Prepare co-occurance matrix
coocur <- ps3 %>%
    #subset_samples(!psyllid_spp == "Atmetocranium_myersi") %>%
    otu_table %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0))

colnames(coocur) <- colnames(coocur) %>%
           str_replace_all(" |-", "_")
rownames(coocur) <- rownames(coocur) %>%
           str_replace_all(" |-", "_")

# H cophenetic distance
h_tree <- pruned.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree) /1e+6) #convert to Mya so integers are small enough for PACO

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none')

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)


# run paco
#paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) 
# Get interaction-specific cophylogenetic contributions using jacknifing
##paco_run <- paco_links(paco_run)
#write_rds(paco_run, "output/cophylogeny/paco_psyllid_microbe_filtered.rds")

# Read in filtered 
paco_run <- readRDS("output/cophylogeny/paco_psyllid_microbe_filtered.rds")
# Read in unfiltered 
#paco_run <- readRDS("output/cophylogeny/paco_psyllid_microbe.rds")
#print overall significance
print(paco_run$gof)

links <- data.frame(joint=names(paco_run$jackknife$mean), values=unname(paco_run$jackknife$mean), upper=unname(paco_run$jackknife$upper)) %>%
  mutate(OTU = joint) %>%
 separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
gg.links <- links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1")) +
  theme_classic()+ 
  theme(axis.text.x = element_blank())

dir.create("output/cophylogeny/psyllid_microbiome")
write_csv(links, "output/cophylogeny/psyllid_microbiome/psyllid_microbiome_links.csv")
write_csv(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_microbiome/microbe_weights.csv")
write_csv(links %>% 
    group_by(Sample_Name) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_microbiome/psyllid_weights.csv")


#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
#PF_run <- parafit(h_dist, s_dist, t(coocur), nperm=99, test.links=TRUE, silent=TRUE)

PF_run <- readRDS("output/cophylogeny/parafit_psyllid_microbe_filtered.rds")

PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite))) %>%
  dplyr::rename(psyllid_spp = `rownames(h_dist)`, OTU = `rownames(s_dist)`) %>%
  mutate(signif = case_when(
    p.F1 < 0.05 ~ 1,
    p.F1 > 0.05 ~ 0
  )) 

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# psyllid_tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links)) #+ geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = Sample_Name)
    )  

## Get values for higher nodes
weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1 + geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# OTU tree
tree2 <- obj[["trees"]][[2]]

# make a clade label list
tax_groups <- as.data.frame(tax_table(ps3)) %>%
  rownames_to_column("label") %>%
  dplyr::mutate(label = label %>% str_replace_all("-", "_") %>%
                  str_replace_all(" ", "_")) %>%
  select(label, genus) %>%
  filter(label %in% tree2$tip.label) %>%
  group_by(genus) 

group_name <- group_keys(tax_groups)  %>%
  mutate(group_name = genus %>% str_remove_all("\\[|\\]"))

cls <- tax_groups %>% 
  group_split() %>% 
  purrr::map(pull, label) %>%
  set_names(group_name$group_name)

newtree <- groupOTU(tree2, cls)
p2 <- ggtree(newtree, ladderize=FALSE, aes(colour=links)) + 
  geom_text2(aes(subset=!isTip, label=group %>% na_if(0)), hjust=0, check_overlap=TRUE)
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = OTU)
    ) 

weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=links)) +  
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links  %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values, upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values, upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup

gg.tangle <- ggplot(tangle, aes(x=tree, y=y, colour=signif)) +
    geom_line(aes(group=assoc),alpha=0.1) +  
  geom_text(data = tangle %>% 
              dplyr::filter(tree=="host") %>%
              group_by(label, y, tree) %>%
              summarise(signif = mean(signif)),
            aes(label=label),stat = 'unique', hjust=1)+
    scale_colour_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = expansion(add=c(0.5,0))) + 
    theme_void() +
    theme(legend.position = "bottom")

whole_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
whole_tanglegram

pdf(file="figs/microbiome_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(whole_tanglegram)
try(dev.off(), silent=TRUE)
  
 # Make plot ranking links, and nodes

gg.para_hlinks <- PF_links %>%
  dplyr::rename(label = psyllid_spp) %>%
  drop_na() %>%
  group_by(label) %>%
  summarise(values = sum(signif)) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="ParaFit Psyllid taxa") + 
  coord_flip()

gg.paco_hlinks <- d1_new %>%
  drop_na() %>%
  group_by(label) %>%
  summarise(values = sum(signif)) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="PACo Psyllid taxa") + 
  coord_flip()

gg.para_slinks <- PF_links %>%
  dplyr::rename(label = OTU) %>%
  left_join(tax_groups) %>%
  drop_na() %>%
  mutate(genus = str_sub(genus, 1, 30)) %>%
  group_by(genus) %>%
  summarise(values = sum(signif)) %>%
  dplyr::rename(label = genus) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
    top_frac(.5) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="ParaFit Top 50% Bacterial genera") + 
  coord_flip()
  
gg.paco_slinks <- d2_new %>%
  left_join(tax_groups) %>%
  drop_na() %>%
  group_by(genus) %>%
  summarise(values = sum(signif)) %>%
  dplyr::rename(label = genus) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  top_frac(.5) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  #theme(axis.text.x = element_text(angle=45, hjust=1)) +
  labs(y = "# Signficant links", x=NULL, title = "PACo Top 50% Bacterial genera") + 
  coord_flip()

gg.psyllid_hlinks <- gg.para_hlinks + gg.paco_hlinks 

gg.psyllid_slinks <- gg.para_slinks + gg.paco_slinks

pdf(file="figs/psyllid_microbiome_host_fit_summary.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.psyllid_hlinks)
try(dev.off(), silent=TRUE)
  
pdf(file="figs/psyllid_microbiome_symbiont_fit_summary.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.psyllid_slinks)
try(dev.off(), silent=TRUE)
  
```

## Psyllid ~ Carsonella 

```{r carsonella cophylo}
#Flag top abundance carsonella by sample
top_carson <- ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  filter(genus=="Candidatus Carsonella") %>%
  group_by(Sample) %>%  
  filter(Abundance > 0) %>%
  top_n(1, wt=Abundance) %>%
  mutate(top = TRUE) 

coocur <- ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  left_join(top_carson) %>%
  filter(genus=="Candidatus Carsonella", top==TRUE) %>%
  dplyr::select(OTU, psyllid_spp, SampleID, Abundance) %>%
  mutate(OTU = OTU %>%
           str_replace_all(" |-", "_")) %>%
  dplyr::group_by(OTU, psyllid_spp) %>%
    summarise(Abundance = sum(Abundance)) %>%
  pivot_wider(id_cols = psyllid_spp,
              names_from = OTU,
              values_from=Abundance,
              values_fill = list(Abundance = 0))  %>%
  column_to_rownames("psyllid_spp") %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0)) 

# H cophenetic distance
h_tree <- pruned.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree)/1e+6 ) ##convert to Mya so integers are small enough for PACO

#alternatively - sqrt(cophenetic(s_tree))
coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none') 

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=99, seed=909, method='quasiswap', symmetric=TRUE)

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using jacknifing
paco_run <- paco_links(paco_run)
links <- data.frame(joint=names(paco_run$jackknife$mean), #losing sv50 here?
                    values=unname(paco_run$jackknife$mean),
                    upper=unname(paco_run$jackknife$upper)) %>%
  mutate(OTU = joint) %>%
 separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1")) +
  theme_classic()+ 
  theme(axis.text.x = element_blank())

dir.create("output/cophylogeny/psyllid_carsonella")
write_csv(links, "output/cophylogeny/psyllid_carsonella/psyllid_carsonella_links.csv")
write_csv(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_carsonella/carsonella_weights.csv")
write_csv(links %>% 
    group_by(Sample_Name) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_carsonella/psyllid_weights.csv")

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")
# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
PF_run <- parafit(h_dist, s_dist,coocur, nperm=99, test.links=TRUE)
PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite)))

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# psyllid_tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links))
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = Sample_Name)
    )  

## Get values for higher nodes
weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1 + geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1")  +
  theme(legend.position = "none")

# OTU tree
tree2 <- obj[["trees"]][[2]]
s_tree <- drop.tip(tree2, setdiff(tree2$tip.label, obj$assoc[,2]))

p2 <- ggtree(tree2 , ladderize=TRUE, aes(colour=links)) 
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = OTU)
    ) 

weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=links)) + 
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links  %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values, upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values, upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup
gg.tangle <- ggplot(tangle, aes(x=tree, y=y, group=assoc, colour=as.factor(signif))) +
    geom_line(alpha=0.8) +  
  geom_text(data = tangle %>% 
              filter(tree=="host"),
            aes(label=label),stat = 'unique', hjust=1)+
  geom_text(data = tangle %>% 
              filter(tree=="microbe"),
            aes(label=label),stat = 'unique', hjust=0)+
    scale_color_manual(values=c("steelblue", "darkorange1")) +
    scale_x_discrete(expand = expansion(add=c(0.8,0.8))) + 
    theme_void() +
    theme(legend.position = "bottom")

gg.carson_tangle <- p1 + gg.tangle + (p2 + scale_x_reverse()) #+ plot_layout(widths = c(2, 1, 2))
gg.carson_tangle 

pdf(file="figs/carsonella_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.carson_tangle)
try(dev.off(), silent=TRUE)
```

## Psyllid ~  hostplant

Tanglegram of all plants and  psyllids!
```{r hostplant-psyllid}
plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.tree  <- drop.tip(plant.tree , "Sophora_microphylla_-kowhai" )

#Prepare co-occurance matrix
coocur <- sample_data(ps2) %>%
  as_tibble() %>%
  dplyr::select(psyllid_spp, hostplant_spp) %>%
  filter(!is.na(psyllid_spp)) %>%
  unique() %>%
  mutate(psyllid_spp = psyllid_spp %>%
           str_replace_all(" |-", "_"),
         hostplant_spp = hostplant_spp %>%
           str_replace_all(" |-", "_"),
         presence = 1
         ) %>%
  pivot_wider(names_from = "hostplant_spp",
              values_from="presence",
              values_fill = list(presence = 0)) %>%
  column_to_rownames("psyllid_spp") %>%
  t()

# H cophenetic distance
h_tree <- plant.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- pruned.tree
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree))

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none') 

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=99, seed=909, method='quasiswap', symmetric=TRUE)

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using leave-one-out jacknifing
paco_run <- paco_links(paco_run, .parallel = TRUE)

# get links
links <- data.frame(joint=names(paco_run$jackknife$mean), values=unname(paco_run$jackknife$mean), upper=unname(paco_run$jackknife$upper)) %>%
 mutate(OTU = joint) %>%
 separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1")) +
  theme_classic()+ 
  theme(axis.text.x = element_blank())

dir.create("output/cophylogeny/psyllid_hostplant")
write_csv(links, "output/cophylogeny/psyllid_hostplant/psyllid_hostplant_links.csv")
write_csv(links %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_hostplant/psyllid_weights.csv")
write_csv(links %>% 
    group_by(hostplant_spp) %>%
    summarise(values = mean(values)), "output/cophylogeny/psyllid_hostplant/hostplant_weights.csv")

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge")

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  #facet_wrap(~psyllid_genus) +
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
PF_run <- parafit(h_dist, s_dist, t(coocur), nperm=999, test.links=TRUE, silent=FALSE)
PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite)))

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# plant tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links)) 
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(hostplant_spp) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = hostplant_spp)
    )

weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1  +
  geom_tippoint(aes(colour=links)) + 
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# psyllid_tree
tree2 <- obj[["trees"]][[2]]
p2 <- ggtree(tree2, ladderize=FALSE, aes(colour=links))
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = psyllid_spp)
    ) 
weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  +
  geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links %>% 
              dplyr::rename(label.x = hostplant_spp, label.y = psyllid_spp)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values,upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values,upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

# Should be able to have y = labels, but arrange b y
tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup
gg.tangle <- ggplot(tangle, aes(x=factor(tree, levels=c("microbe", "host")), y=y, group=assoc, colour=as.factor(signif))) +
    geom_line(alpha=0.8) +  
  geom_text(data = tangle %>% 
              filter(tree=="host"),
            aes(label=label),stat = 'unique', hjust=0)+
  geom_text(data = tangle %>% 
              filter(tree=="microbe"),
            aes(label=label),stat = 'unique', hjust=1)+
    scale_color_manual(values=c("steelblue", "darkorange1"), name = "Significant") +
    scale_x_discrete(expand = expansion(add=c(0.8,0.8))) + 
    theme_void() +
    theme(legend.position = "bottom")

gg.plant_tangle <- p2 + gg.tangle + (p1 + scale_x_reverse()) #+ plot_layout(widths = c(2, 1, 2))
gg.plant_tangle

pdf(file="figs/plant_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.plant_tangle)
try(dev.off(), silent=TRUE)
```

# Trioza

## Trioza ~ Microbe

```{r trioza microbe}
#Prepare co-occurance matrix
coocur <- ps3 %>%
    subset_samples(psyllid_genus == "Trioza") %>%
    filter_taxa(function(x) mean(x) > 0, TRUE) %>%
    otu_table %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0))

colnames(coocur) <- colnames(coocur) %>%
           str_replace_all(" |-", "_")
rownames(coocur) <- rownames(coocur) %>% 
           str_replace_all(" |-", "_")

# H cophenetic distance
h_tree <- pruned.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree) /1e+6) #convert to Mya so integers are small enough for PACO

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]
# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none')

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
#paco_run <- PACo(D, nperm=9, seed=909, method='quasiswap', symmetric=TRUE) 
# Get interaction-specific cophylogenetic contributions using jacknifing
##paco_run <- paco_links(paco_run)
#write_rds(paco_run, "output/cophylogeny/paco_trioza_microbe_filtered.rds")

# Read in filtered 
paco_run <- readRDS("output/cophylogeny/paco_trioza_microbe_filtered.rds")
# Read in unfiltered 
#paco_run <- readRDS("output/cophylogeny/paco_trioza_microbe.rds")

#print overall significance
print(paco_run$gof)


links <- data.frame(joint=names(paco_run$jackknife$mean), values=unname(paco_run$jackknife$mean), upper=unname(paco_run$jackknife$upper)) %>%
  mutate(OTU = joint) %>%
 separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1"), name = "Significant") +
  theme_classic()+ 
  theme(axis.text.x = element_blank())

dir.create("output/cophylogeny/trioza_microbiome")
write_csv(links, "output/cophylogeny/trioza_microbiome/psyllid_microbiome_links.csv")
write_csv(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_microbiome/microbe_weights.csv")
write_csv(links %>% 
    group_by(Sample_Name) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_microbiome/psyllid_weights.csv")


#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
#PF_run <- parafit(h_dist, s_dist, t(coocur), nperm=99, test.links=TRUE, silent=TRUE)
#write_rds(PF_run, "output/cophylogeny/parafit_trioza_microbe_filtered.rds")

# Read in filtered
PF_run <- readRDS("output/cophylogeny/parafit_trioza_microbe_filtered.rds")
# Read in unfiltered
#PF_run <- readRDS("output/cophylogeny/parafit_trioza_microbe_unfiltered.rds")

PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite))) %>%
  dplyr::rename(psyllid_spp = `rownames(h_dist)`, OTU = `rownames(s_dist)`) %>%
  mutate(signif = case_when(
    p.F1 < 0.05 ~ 1,
    p.F1 > 0.05 ~ 0
  )) 

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# psyllid_tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links))
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = Sample_Name)
    )  

## Get values for higher nodes
weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1 + geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1") +
    theme(legend.position = "none")

# OTU tree
tree2 <- obj[["trees"]][[2]]

# make a clade label list
tax_groups <- as.data.frame(tax_table(ps3)) %>%
  rownames_to_column("label") %>%
  dplyr::mutate(label = label %>% str_replace_all("-", "_") %>%
                  str_replace_all(" ", "_")) %>%
  select(label, genus) %>%
  filter(label %in% tree2$tip.label) %>%
  group_by(genus) 

group_name <- group_keys(tax_groups)  %>%
  mutate(group_name = genus %>% str_remove_all("\\[|\\]"))

cls <- tax_groups %>% 
  group_split() %>% 
  purrr::map(pull, label) %>%
  set_names(group_name$group_name)

newtree <- groupOTU(tree2, cls)
p2 <- ggtree(newtree, ladderize=TRUE, aes(colour=links)) + 
  geom_text2(aes(subset=!isTip, label=group %>% na_if(0)), hjust=0, check_overlap=TRUE)
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = OTU)
    ) 

weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  + geom_tippoint(aes(colour=links)) +  
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")


# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values, upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y,  values, upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup

gg.tangle <- ggplot(tangle, aes(x=tree, y=y, colour=signif)) +
    geom_line(aes(group=assoc),alpha=0.1) +  
  geom_text(data = tangle %>% 
              dplyr::filter(tree=="host") %>%
              group_by(label, y, tree) %>%
              summarise(signif = mean(signif)),
            aes(label=label),stat = 'unique', hjust=1)+
    scale_colour_gradient(low="steelblue", high="darkorange1") +
    scale_x_discrete(expand = expansion(add=c(0.5,0))) + 
    theme_void() +
    theme(legend.position = "bottom")

trioza_tanglegram <- p1 + gg.tangle + (p2 + scale_x_reverse())
trioza_tanglegram

pdf(file="figs/trioza_microbe_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(trioza_tanglegram)
try(dev.off(), silent=TRUE)
  
 # Make plot ranking links, and nodes

gg.para_hlinks <- PF_links %>%
  dplyr::rename(label = psyllid_spp) %>%
  drop_na() %>%
  group_by(label) %>%
  summarise(values = sum(signif)) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="ParaFit Psyllid taxa") + 
  coord_flip()

gg.paco_hlinks <- d1_new %>%
  drop_na() %>%
  group_by(label) %>%
  summarise(values = sum(signif)) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="PACo Psyllid taxa") + 
  coord_flip()

gg.para_slinks <- PF_links %>%
  dplyr::rename(label = OTU) %>%
  left_join(tax_groups) %>%
  drop_na() %>%
  group_by(genus) %>%
  summarise(values = sum(signif)) %>%
  dplyr::rename(label = genus) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum ))%>%
  arrange(-values) %>%
    top_frac(.5) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  labs(y = "# Signficant links", x=NULL, title ="ParaFit Top 50% Bacterial genera") + 
  coord_flip()
  
gg.paco_slinks <- d2_new %>%
  left_join(tax_groups) %>%
  drop_na() %>%
  group_by(genus) %>%
  summarise(values = sum(signif)) %>%
  dplyr::rename(label = genus) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  top_frac(.5) %>%
  ggplot(aes(x=label, y=values, fill=values)) +
  geom_col(show.legend = FALSE) +  
  scale_fill_gradient(low="steelblue", high="darkorange1") +
  theme_classic()+ 
  #theme(axis.text.x = element_text(angle=45, hjust=1)) +
  labs(y = "# Signficant links", x=NULL, title = "PACo Top 50% Bacterial genera") + 
  coord_flip()

gg.trioza_hlinks <- gg.para_hlinks + gg.paco_hlinks 

gg.trioza_slinks <- gg.para_slinks + gg.paco_slinks

pdf(file="figs/trioza_microbiome_host_fit_summary.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.trioza_hlinks)
try(dev.off(), silent=TRUE)
  
pdf(file="figs/trioza_microbiome_symbiont_fit_summary.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.trioza_slinks)
try(dev.off(), silent=TRUE)
```

## Trioza ~ Carsonella

```{r Trioza carsonella}
#Flag top abundance carsonella by sample
top_carson <- ps2 %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  filter(genus=="Candidatus Carsonella") %>%
  group_by(Sample) %>%  
  filter(Abundance > 0) %>%
  top_n(1, wt=Abundance) %>%
  mutate(top = TRUE) 

coocur <- ps2 %>%
  subset_samples(psyllid_genus == "Trioza") %>%
  filter_taxa(function(x) mean(x) > 0, TRUE) %>%
  transform_sample_counts(function (x) x/sum(x)) %>%
  speedyseq::psmelt() %>%
  left_join(top_carson) %>%
  filter(genus=="Candidatus Carsonella", top==TRUE) %>%
  dplyr::select(OTU, psyllid_spp, SampleID, Abundance) %>%
  mutate(OTU = OTU %>%
           str_replace_all(" |-", "_")) %>%
  dplyr::group_by(OTU, psyllid_spp) %>%
    summarise(Abundance = sum(Abundance)) %>%
  pivot_wider(id_cols = psyllid_spp,
              names_from = OTU,
              values_from=Abundance,
              values_fill = list(Abundance = 0))  %>%
  column_to_rownames("psyllid_spp") %>%
    as.matrix() %>% 
  apply(2, function(x) ifelse(x > 0, 1, 0)) 

# H cophenetic distance
h_tree <- pruned.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- phy_tree(ps3)
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree)/1e+6 ) ##convert to Mya so integers are small enough for PACO

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]

# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none') 

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=99, seed=909, method='quasiswap', symmetric=TRUE)

#print overall significance
print(paco_run$gof)

# Get interaction-specific cophylogenetic contributions using jacknifing
paco_run <- paco_links(paco_run)
links <- data.frame(joint=names(paco_run$jackknife$mean), #losing sv50 here?
                    values=unname(paco_run$jackknife$mean), upper=unname(paco_run$jackknife$upper)) %>%
  mutate(OTU = joint) %>%
 separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1")) +
  theme_classic()+ 
  theme(axis.text.x = element_blank())

dir.create("output/cophylogeny/trioza_carsonella")
write_csv(links, "output/cophylogeny/trioza_carsonella/trioza_carsonella_links.csv")
write_csv(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_carsonella/carsonella_weights.csv")
write_csv(links %>% 
    group_by(Sample_Name) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_carsonella/psyllid_weights.csv")

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")
# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("Sample_Name", "OTU"), sep="-", extra="merge") 

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
PF_run <- parafit(h_dist, s_dist, t(coocur), nperm=999, test.links=TRUE, silent=FALSE)
PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite)))

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# psyllid_tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links)) #+ geom_tiplab()
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(Sample_Name) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = Sample_Name)
    )  

## Get values for higher nodes
weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1 + 
  geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")


# OTU tree
tree2 <- obj[["trees"]][[2]]

p2 <- ggtree(tree2 , ladderize=TRUE, aes(colour=links)) 
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(OTU) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = OTU)
    ) 

weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  + 
  geom_tippoint(aes(colour=links)) +  
  scale_color_gradient(low="steelblue", high="darkorange1")+
  theme(legend.position = "none")


# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links  %>% 
              dplyr::rename(label.x = Sample_Name, label.y = OTU)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values, upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values, upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup
gg.tangle <- ggplot(tangle, aes(x=tree, y=y, group=assoc, colour=as.factor(signif))) +
    geom_line(alpha=0.8) +  
  geom_text(data = tangle %>% 
              filter(tree=="host"),
            aes(label=label),stat = 'unique', hjust=1)+
  geom_text(data = tangle %>% 
              filter(tree=="microbe"),
            aes(label=label),stat = 'unique', hjust=0)+
    scale_color_manual(values=c("steelblue", "darkorange1")) +
    scale_x_discrete(expand = expansion(add=c(0.8,0.8))) + 
    theme_void() +
    theme(legend.position = "bottom")

gg.trioza_carson_tangle <- p1 + gg.tangle + (p2 + scale_x_reverse()) 
gg.trioza_carson_tangle

pdf(file="figs/trioza_carsonella_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.trioza_carson_tangle)
try(dev.off(), silent=TRUE)
```

## Trioza ~ hostplant

```{r trioza hostplant psyllid}
plant.tree <- read.tree("sample_data/plant_tree.nwk")
plant.tree  <- drop.tip(plant.tree , "Sophora_microphylla_-kowhai" )

#Prepare co-occurance matrix
coocur <- sample_data(ps2) %>%
  as_tibble() %>%
  filter(psyllid_genus == "Trioza") %>%
  dplyr::select(psyllid_spp, hostplant_spp) %>%
  filter(!is.na(psyllid_spp)) %>%
  unique() %>%
  mutate(psyllid_spp = psyllid_spp %>%
           str_replace_all(" |-", "_"),
         hostplant_spp = hostplant_spp %>%
           str_replace_all(" |-", "_"),
         presence = 1
         ) %>%
  pivot_wider(names_from = "hostplant_spp",
              values_from="presence",
              values_fill = list(presence = 0)) %>%
  column_to_rownames("psyllid_spp") %>%
  t()

# H cophenetic distance
h_tree <- plant.tree
h_tree$tip.label <- h_tree$tip.label %>%
           str_replace_all(" |-", "_")
h_tree <- drop.tip(h_tree, setdiff(h_tree$tip.label, rownames(coocur)))
h_dist <- sqrt(cophenetic(h_tree))

# P cophenetic distance
s_tree <- pruned.tree
s_tree$tip.label <- s_tree$tip.label %>%
           str_replace_all(" |-", "_")
s_tree <- drop.tip(s_tree, setdiff(s_tree$tip.label, colnames(coocur)))
s_dist <- sqrt(cophenetic(s_tree))

coocur <- coocur[h_tree$tip.label, s_tree$tip.label]
# prepare paco data
D <- prepare_paco_data(H=h_dist, P=s_dist, HP=coocur)
# Add pcord
D <- add_pcoord(D, correction='none')  

p_host <- ggplot(D$H_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2)) +
    geom_point() +
    theme_bw() +
  ggtitle("H PCA")

p_para <- ggplot(D$P_PCo[,c('Axis.1', 'Axis.2')] %>% as.data.frame, aes(Axis.1, Axis.2))  +
    geom_point() +
    theme_bw()+
  ggtitle("P PCA")
    
plot(p_host + p_para)

# run paco
paco_run <- PACo(D, nperm=99, seed=909, method='quasiswap', symmetric=TRUE) #Symetric - is one meant to track the evolution of another?

#print overall significance
print(paco_run$gof)

# Procrustes diagnostic plots
plot(paco_run$proc)
plot(paco_run$proc, kind=2)


# Get interaction-specific cophylogenetic contributions using jacknifing
paco_run <- paco_links(paco_run)
links <- data.frame(joint=names(paco_run$jackknife$mean), values=unname(paco_run$jackknife$mean), upper=unname(paco_run$jackknife$upper)) %>%
 mutate(OTU = joint) %>%
 separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge") %>%
  mutate(signif = case_when(
    upper < mean(.$values) ~ 1,
    upper > mean(.$values) ~ 0
  ))
# Plot links
links %>%
  dplyr::rename(label = joint) %>%
  mutate(label = as.factor(label),
         label=fct_reorder(label, values, sum))%>%
  arrange(-values) %>%
  ggplot(aes(x=label, y=values, colour=as.factor(signif))) +
  geom_point(show.legend = FALSE) +
  geom_errorbar(aes(ymin=values, ymax=upper)) +
  geom_hline(yintercept = mean(links$values)) +
  scale_color_manual(values=c("steelblue", "darkorange1")) +
  theme_classic()+ 
  theme(axis.text.x = element_text(angle=45, hjust=1))

dir.create("output/cophylogeny/trioza_hostplant")
write_csv(links, "output/cophylogeny/trioza_hostplant/psyllid_hostplant_links.csv")
write_csv(links %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_hostplant/psyllid_weights.csv")
write_csv(links %>% 
    group_by(hostplant_spp) %>%
    summarise(values = mean(values)), "output/cophylogeny/trioza_hostplant/hostplant_weights.csv")

#Get observed residuals of Procrustean superimposition 
paco_residuals <- residuals_paco(paco_run$proc, type = "interaction")

# Visualise residuals
res <- data.frame(OTU=names(paco_residuals), values=unname(paco_residuals)) %>%
  separate(OTU, into=c("hostplant_spp", "psyllid_spp"), sep="-", extra="merge")

ggplot(res, aes(x=values))+
  geom_density(fill='grey70')+
  theme_bw()+
  xlab('Procrustes residuals')+
  ylab('Frequency')

# Parafit run
PF_run <- parafit(h_dist, s_dist, t(coocur), nperm=999, test.links=TRUE, silent=TRUE)
PF_run$ParaFitGlobal
PF_run$p.global

PF_links <- as.data.frame(PF_run$link.table) %>%
  left_join(as.data.frame(rownames(h_dist)) %>%
              rownames_to_column("Host") %>%
              mutate(Host = as.numeric(Host))) %>%
  left_join(as.data.frame(rownames(s_dist)) %>%
              rownames_to_column("Parasite") %>%
              mutate(Parasite = as.numeric(Parasite)))

# Cophyloplot
coocur.lut <- which(coocur ==1, arr.ind=TRUE)
assoc <- cbind(rownames(coocur)[coocur.lut[,1]], colnames(coocur)[coocur.lut[,2]])

# Rotate the nodes using phytools
library(phytools)
obj <- cophylo(tr1=h_tree, tr2=s_tree, assoc=assoc, rotate=TRUE) 

# Extract the goods for ggtree
# plant tree
tree1 <- obj[["trees"]][[1]]
p1 <- ggtree(tree1, ladderize=FALSE, aes(colour=links)) 
weights_p1 <- p1$data %>%
  left_join(links %>%
  group_by(hostplant_spp) %>%
  summarise(values = mean(signif)) %>%
    dplyr::rename(label = hostplant_spp)
    )

weights_p1 <- weights_p1 %>%
  left_join(data.frame(node=weights_p1$node, links = weights_p1$node %>% purrr::map_dbl(average_descendants, tree=tree1, df=weights_p1)))

p1 <- p1 %<+% weights_p1  + 
  geom_tippoint(aes(colour=links)) +
  scale_color_gradient(low="steelblue", high="darkorange1") +
  theme(legend.position = "none")

# psyllid_tree
tree2 <- obj[["trees"]][[2]]
p2 <- ggtree(tree2, ladderize=FALSE, aes(colour=links))
weights_p2 <- p2$data %>%
  left_join(links %>% 
    group_by(psyllid_spp) %>%
    summarise(values = mean(signif)) %>%
    dplyr::rename(label = psyllid_spp)
    ) 
weights_p2 <- weights_p2 %>%
  left_join(data.frame(node=weights_p2$node, links = weights_p2$node %>% purrr::map_dbl(average_descendants, tree=tree2, df=weights_p2)))

p2 <- p2 %<+% weights_p2  + 
  geom_tippoint(aes(colour=links)) +  
  scale_color_gradient(low="steelblue", high="darkorange1")+
  theme(legend.position = "none")

# Tanglegram
dd <- obj$assoc %>%
  as_data_frame() %>%
  magrittr::set_colnames(c("label.x", "label.y")) %>%
  left_join(links %>% 
              dplyr::rename(label.x = hostplant_spp, label.y = psyllid_spp)) %>%
  left_join(p1$data %>% dplyr::select(label, y) %>% dplyr::rename(label.x = label), by="label.x") %>%
  left_join(p2$data %>% dplyr::select(label, y) %>% dplyr::rename(label.y = label), by="label.y") %>%
  rownames_to_column("assoc")

d1_new <- dd %>% 
  dplyr::select(assoc, label.x, y.x, values, upper, signif) %>%
  dplyr::rename(label = label.x, y = y.x)%>%
  mutate(tree = "host") 
  
d2_new <- dd %>% 
  dplyr::select(assoc, label.y, y.y, values, upper, signif) %>%
  dplyr::rename(label = label.y, y = y.y) %>%
  mutate(tree = "microbe")

# Should be able to have y = labels, but arrange b y

tangle <- bind_rows(d1_new, d2_new) %>%
  filter(!is.na(label))%>% 
  group_by(tree) %>%
  mutate(y = y / max(y)) %>%
  ungroup
gg.tangle <- ggplot(tangle, aes(x=factor(tree, levels=c("microbe", "host")), y=y, group=assoc, colour=as.factor(signif))) +
    geom_line(alpha=0.8) +  
  geom_text(data = tangle %>% 
              filter(tree=="host"),
            aes(label=label),stat = 'unique', hjust=0)+
  geom_text(data = tangle %>% 
              filter(tree=="microbe"),
            aes(label=label),stat = 'unique', hjust=1)+
    scale_color_manual(values=c("steelblue", "darkorange1"), name = "Significant") +
    scale_x_discrete(expand = expansion(add=c(0.8,0.8))) + 
    theme_void() +
    theme(legend.position = "bottom")

gg.triozid_tangle <- p2 + gg.tangle + (p1 + scale_x_reverse()) 
gg.triozid_tangle

pdf(file="figs/trioza_plant_tanglegram.pdf",  width = 8, height = 11, paper="a4")
  plot(gg.triozid_tangle)
try(dev.off(), silent=TRUE)
```

# Map of colleciton locations

```{r Collection map}
#Plot on map to confirm points
library(rgeos)
library(maptools)
envData <- sample_data(ps2) %>%
  as_data_frame() %>%
  dplyr::select(Sample_Name,psyllid_spp, lat, long) %>%
  tibble::column_to_rownames("Sample_Name") %>%
  drop_na()

xlim <- c(165,180)
ylim <- c(-50,-30)

nz <-map("worldHires", "New Zealand", fill=TRUE, xlim=xlim,
  ylim=ylim) #, mar=c(0,0,0,0)

p1 <- ggtree(pruned.tree, ladderize=TRUE)
map_data <- p1$data%>%
  left_join(envData %>% dplyr::mutate(label =psyllid_spp ))


col <- colorRampPalette(brewer.pal(12, "Paired"))(length(unique(map_data$psyllid_spp)))

gg.nzmap <- ggplot(fortify(nz), aes(y=lat, x=long, group=group)) + 
  geom_polygon(fill="lightgrey", color="#7f7f7f") +
  geom_point(data=map_data, aes(x=long, y=lat, color=psyllid_spp), alpha=.5, size=3, inherit.aes = FALSE) + 
    geom_segment(data=map_data %>%
  mutate(y = scales::rescale(y, to = ylim )), aes(x=min(xlim), y=y, xend= long, yend=lat, color=psyllid_spp), alpha=.2, inherit.aes = FALSE) +
    theme_classic() +
    coord_fixed(ylim =ylim, xlim=xlim) +
    scale_x_continuous(expand = c(0,0)) + 
    scale_colour_manual(values=col) +
    theme(legend.position = "none") +
    scale_y_continuous(position = "right") +
  xlab("Longitude") + 
  ylab("Lattitude")
gg.nzmap 

#print(gg.nzmap, vp = viewport(width = .7, height = .7, angle = 35))
p1 <- p1 %<+% map_data +
  geom_tiplab(align=TRUE, aes(color=psyllid_spp), offset=0.1, hjust=1) +
  scale_x_continuous(expand=c(0, 0))+ 
  scale_colour_manual(values=col) 

gg.phylomap <- p1 + gg.nzmap

pdf(file="figs/phylomap.pdf",  width = 15, height = 15, paper="a4r")
  plot(gg.phylomap)
try(dev.off(), silent=TRUE)
```

# Sessioninfo

```{r sessioninfo, eval=TRUE}
devtools::session_info()
```
